---
title: "Reliable and extensive analysis of Proteomics data with package `VisomX`"
author: "Nicolas T. Wirth"
date: "2022-01-31"
mainfont: Arial
output:
  prettydoc::html_pretty:
    fig_caption: yes
    toc: true
    theme: "cayman"
    highlight: vignette
vignette: >
  %\VignetteIndexEntry{Reliable and extensive analysis of Proteomics data with package `VisomX`}
  %\VignetteEncoding{UTF-8}
  %\VignetteEngine{knitr::rmarkdown}
csl: https://raw.githubusercontent.com/citation-style-language/styles/master/research-institute-for-nature-and-forest.csl
link-citations: true
lang: en-US
zotero: true
bibliography: references.bib
editor_options: 
  markdown: 
    wrap: 72
---

```{=latex}
\definecolor{codegray}{HTML}{efefef}
\let\textttOrig\texttt
\renewcommand{\texttt}[1]{\textttOrig{\colorbox{codegray}{#1}}}
```
<p style="text-align:center;">

<img src="https://user-images.githubusercontent.com/43075653/166063689-45b4e34e-7129-42c3-b109-3a34afdc24e2.png" class="center" width="433" height="178"/>

</p>

# Introduction

The aim of this `VisomX`'s proteomics section is to provide an
integrated analysis workflow for robust and reproducible analysis of
mass spectrometry proteomics data for differential protein expression or
differential enrichment. This section is build on the foundation of the
R package `DEP` [@zhang2018]. `VisomX` was designed to be usable with
minimal prior knowledge of the R programming language or programming in
general. You will need to be familiar with the idea of running commands
from a console or writing basic scripts. For R beginners,
[this](https://moderndive.netlify.app/1-getting-started.html) is a great
starting point, there are some good resources
[here](https://education.rstudio.com/learn/beginner/) and we suggest
using the [RStudio
application](https://posit.co/products/open-source/rstudio/). It
provides an environment for writing and running R code.

With consideration for `R` novices, `VisomX` establishes a framework in
which a complete, detailed proteomics data analysis can be performed in
two simple steps:

1.  *Read data* from a tabular input with protein abundances, as
    generated by quantitative analysis software of raw mass spectrometry
    data

2.  *Run workflow*, including missing value imputation, normalization,
    log2 transformation, statistical analysis, differential expression
    analysis, pathway enrichment analysis, and rendering of a report
    that summarizes the results.

All computational results of a workflow are stored in a data container
(list) and can be visualized by passing them to a set of dedicated
plotting functions.

\pagebreak

# Installation

Install the most current version with package `devtools`:

``` r
install.packages("devtools")
library(devtools)

install_github("NicWir/VisomX", dependencies = TRUE)
```

Load the package:

```{r echo = T, eval = F, results = "hide"}
# Load the VisomX package
library(VisomX)
```

```{r echo = F, eval = T, include = F, results = "hide"}
# Load the VisomX package
suppressPackageStartupMessages(library(VisomX))
```

# The Test Dataset

The proteomics test dataset was generated by culturing *Pseudomonas
putida* KT2440 in 500-mL shaken flasks filled with 50 mL media at 30°C
and 250 rpm. The media were either synthetic de Bont medium supplemented
with 20 mM glucose or 30 mM acetate, or LB medium. Samples were taken in
the mid-exponential phase, and cells were harvested by centrifugation at
17,000 x g for 2 min at 4°C. After removal of the supernatant, cell
pellets were frozen and kept at −80°C until proteomics analysis was
performed as described previously [@rennig2019]. No normalization of
protein abundances was performed prior to this analysis.

The path to the example dataset can be extracted with:

```{r echo = T, eval = T, include = T, results = "asis"}
system.file("prot_KT2440_glc_ac_LB.txt", package = "VisomX")

```

## Data Format

`VisomX` requires a tabular input, either in the form of an R dataframe,
or stored in an *XLS*, *XLSX*, *CSV*, *TSV*, or *TXT* file. Three types
of columns are required:

1.  A column containing protein identifiers. These can be Uniprot
    identifiers, Ensembl Gene IDs, Entrez Gene IDs, etc.

2.  A column containing names. Gene Symbols are commonly used for this
    purpose.

3.  Columns containing protein abundances of the different samples in
    the experiment. These columns should have a common prefix (e.g.,
    *abundance.*) used for their identification within the table.
    Furthermore, replicates are indicated by an underscore followed by a
    number. For example, *abundance.KT2440_Glc_1*,
    *abundance.KT2440_Glc_2*, *abundance.KT2440_Glc_3*, and
    *abundance.KT2440_Glc_4* indicate four replicates for the condition
    *KT2440_Glc*.

4.  *Optional*: Columns containing *+* and *-* indicating contaminant
    proteins and decoy database hits that should be filtered out.

# Run a Complete Proteomics Analysis Workflow

## Read Data

The function `prot.read_data()` accepts either an R dataframe or the
path to a table file with extension '.xlsx', '.xls', '.csv', '.tsv', or
'.txt'. It creates a `SummarizedExperiment` object and performs feature
filtering based on indicated contaminants or unspecific identification,
a defined relative standard deviation threshold, or the presence of
missing values. For details, run `help("prot.read_data", "VisomX")`.
Here, we filter proteins that are not present in at least 6 out of 9
samples.

```{r include = T, eval = T}
# Read data, incl. filtering
data <- prot.read_data(
          data = system.file("prot_KT2440_glc_ac_LB.txt", package = "VisomX"),
          pfx = "abundance.",
          id = "Ensembl Gene ID",
          name = "Gene Symbol",
          filt_type = "fraction",
          filt_min = 0.66)
```

## Run Workflow

The wrapper function `prot.workflow()` runs a complete analysis workflow
and exports the results as tabular TXT files as well as (optionally) a
report in PDF and HTML format. The following settings usually need
closer consideration:

-   We use the *"SampMin"* method for missing value imputation
    (`imp_fun`), which has been shown to be among the most reliable
    methods for both missing-not-at-random (MNAR) and
    missing-completely-at-random types of missing values [@liu2021].

-   For the types of comparisons to perform (`type`), we choose *"all"*
    to cover all possible contrasts. Alternatively, one can compare
    every condition with a single control condition by choosing
    `type = 'control'` and defining the control condition with
    `control`. If only very specific contrasts are of interest,
    `type = 'manual'` can be chosen and the desired contrasts can be
    defined as a vector of strings in the form
    `contrast = c('ConditionA_vs_ConditionB', ConditionC_vs_ConditionD')`.
    The `_vs_` substring is hereby essential!

-   We set the significance threshold for adjusted p values (`alpha`) at
    0.05. `VisomX` tests for differential expression of proteins based
    on protein-wise linear models and\
    empirical Bayes statistics using `limma`. False Discovery Rates are
    estimated using `fdrtool`.

-   We define 1 as the *relevance threshold* for log2 fold changes
    (`lfc`), corresponding to fold changes of 2 and 0.5, to be
    considered significant.

-   We choose `pathway_enrichment = TRUE` to perform pathway enrichment
    analysis. We test for enriched KEGG pathways by setting
    `pathway_kegg = TRUE` and defining a `kegg_organism`. We also test
    for enriched pathways extracted from the BioCyc database as a
    tab-delimited TXT file included in `VisomX`. To test against a set
    of pathways other than those listed in KEGG, we provide a dataframe
    object with columns *'Pathway'* and *'Accession'*. **Important**:
    The names or IDs used for proteins/genes listed under 'Accession'
    need to correspond to the IDs initially used to create the
    SummarizedExperiment with `prot.read_data()`.

-   Lastly, we choose to have a report generated and exported into a
    subfolder of the current working directory.

```{r include = T, eval = T}
# Read TXT file with BioCyc pathways for P. putida KT2440
custom_df <-
  read.table(
    system.file("BioCyc_pathways_KT2440.txt", package = "VisomX"),
    sep = "\t",
    header = T,
    stringsAsFactors = F,
    fill = T,
    na.strings = "",
    quote = "",
    comment.char = "",
    check.names = F
  )

# Run workflow and export results
results <- prot.workflow(se = data,
        imp_fun = "SampMin",
        type = "all",
        alpha = 0.05,
        lfc = 1,
        pathway_enrichment = TRUE,
        pathway_kegg = TRUE,
        kegg_organism = "ppu",
        custom_pathways = custom_df,
        report = TRUE,
        report.dir = "Report")
```

# References

::: {#refs}
:::
