---
title: "Reliable and extensive analysis of transcriptomics data with package `VisomX`"
author: "Nicolas T. Wirth"
date: "2022-01-31"
mainfont: Arial
output:
  prettydoc::html_pretty:
    fig_caption: yes
    toc: true
    theme: "cayman"
    highlight: vignette
vignette: >
  %\VignetteIndexEntry{Reliable and extensive analysis of transcriptomics data with package `VisomX`}
  %\VignetteEncoding{UTF-8}
  %\VignetteEngine{knitr::rmarkdown}
csl: https://raw.githubusercontent.com/citation-style-language/styles/master/research-institute-for-nature-and-forest.csl
link-citations: true
lang: en-US
zotero: true
bibliography: references.bib
---

```{=latex}
\definecolor{codegray}{HTML}{efefef}
\let\textttOrig\texttt
\renewcommand{\texttt}[1]{\textttOrig{\colorbox{codegray}{#1}}}
```
<p style="text-align:center;">

<img src="https://user-images.githubusercontent.com/43075653/166063689-45b4e34e-7129-42c3-b109-3a34afdc24e2.png" class="center" width="433" height="178"/>

</p>

# Introduction

The aim of the VisomX's transcriptomics section is to provide an integrated analysis workflow for robust and reproducible analysis of RNAseq data for differential gene expression. This section is built on the foundation of the R package DESeq2 [@love2014]. VisomX was designed to be user-friendly, requiring minimal prior knowledge of the R programming language or programming in general. However, a basic understanding of running commands from a console or writing simple scripts is necessary. For those new to R, [this](https://moderndive.netlify.app/1-getting-started.html) website is a great starting point, there are some good resources [here](https://education.rstudio.com/learn/beginner/) and I suggest using the [RStudio application](https://posit.co/products/open-source/rstudio/). It provides an environment for writing and running R code.

With consideration for `R` novices, `VisomX` establishes a framework in which a complete, detailed RNASeq data analysis can be performed in two simple steps:

1.  *Read data* from one or several tabular inputs with gene counts, assign new names and apply feature filtering.

2.  *Run workflow*, including missing value imputation, normalization, log2 transformation, statistical analysis, differential expression analysis, log2 fold change shrinkage, pathway enrichment analysis, and rendering of a report that summarizes the results.

All computational results of a workflow are stored in a data container (list) and can be visualized by passing them to a set of dedicated plotting functions.

\pagebreak

# Installation

Install the most current version with package `devtools`:

``` r
install.packages("devtools")
library(devtools)

install_github("NicWir/VisomX", dependencies = TRUE)
```

Load the package:

```{r library, echo = T, eval = F, results = "hide"}
# Load the VisomX package
library(VisomX)
```

```{r libraryCode, echo = F, eval = T, include = F, results = "hide"}
# Load the VisomX package
suppressPackageStartupMessages(library(VisomX))
suppressPackageStartupMessages(library(ggplot2))
```



# References

::: {#refs}
:::
