---
title: "Reliable and extensive analysis of Proteomics data with package `VisomX`"
author: "Nicolas T. Wirth"
date: "2022-01-31"
mainfont: Arial
output:
  prettydoc::html_pretty:
    fig_caption: yes
    toc: true
    theme: "cayman"
    highlight: vignette
vignette: >
  %\VignetteIndexEntry{Reliable and extensive analysis of Proteomics data with package `VisomX`}
  %\VignetteEncoding{UTF-8}
  %\VignetteEngine{knitr::rmarkdown}
csl: https://raw.githubusercontent.com/citation-style-language/styles/master/research-institute-for-nature-and-forest.csl
link-citations: true
lang: en-US
zotero: true
bibliography: references.bib
editor_options: 
  markdown: 
    wrap: 72
---

```{=latex}
\definecolor{codegray}{HTML}{efefef}
\let\textttOrig\texttt
\renewcommand{\texttt}[1]{\textttOrig{\colorbox{codegray}{#1}}}
```
<p style="text-align:center;">

<img src="https://user-images.githubusercontent.com/43075653/166063689-45b4e34e-7129-42c3-b109-3a34afdc24e2.png" class="center" width="433" height="178"/>

</p>

# Introduction

The aim of this `VisomX`'s proteomics section is to provide an
integrated analysis workflow for robust and reproducible analysis of
mass spectrometry proteomics data for differential protein expression or
differential enrichment. This section is build on the foundation of the
R package `DEP` [@zhang2018]. `VisomX` was designed to be usable with
minimal prior knowledge of the R programming language or programming in
general. You will need to be familiar with the idea of running commands
from a console or writing basic scripts. For R beginners,
[this](https://moderndive.netlify.app/1-getting-started.html) is a great
starting point, there are some good resources
[here](https://education.rstudio.com/learn/beginner/) and we suggest
using the [RStudio
application](https://posit.co/products/open-source/rstudio/). It
provides an environment for writing and running R code.

With consideration for `R` novices, `VisomX` establishes a framework in
which a complete, detailed proteomics data analysis can be performed in
two simple steps:

1.  *Read data* from a tabular input with protein abundances, as
    generated by quantitative analysis software of raw mass spectrometry
    data

2.  *Run workflow*, including missing value imputation, normalization,
    log2 transformation, statistical analysis, differential expression
    analysis, pathway enrichment analysis, and rendering of a report
    that summarizes the results.

All computational results of a workflow are stored in a data container
(list) and can be visualized by passing them to a set of dedicated
plotting functions.

\pagebreak

# Installation

Install the most current version with package `devtools`:

``` r
install.packages("devtools")
library(devtools)

install_github("NicWir/VisomX", dependencies = TRUE)
```

# The Test Dataset

The proteomics test dataset was generated by culturing *Pseudomonas
putida* KT2440 in 500-mL shaken flasks filled with 50 mL media at 30°C
and 250 rpm. The media were either synthetic de Bont medium supplemented
with 20 mM glucose or 30 mM acetate, or LB medium. Samples were taken in
the mid-exponential phase, and cells were harvested by centrifugation at
17,000 x g for 2 min at 4°C. After removal of the supernatant, cell
pellets were frozen and kept at −80°C until proteomics analysis was
performed as described previously [@rennig2019]. No normalization of
protein abundances was performed prior to this analysis.

The path to the example dataset can be extracted with:

``` r
system.file("prot_KT2440_glc_ac_LB.txt", package = "VisomX")
```

## Data Format

`VisomX` requires a tabular input, either in the form of an R dataframe, or stored in an *XLS*, *XLSX*, *CSV*, *TSV*, or *TXT* file.
Three types of columns are required:

1.  A column containing protein identifiers. These can be Uniprot identifiers, Ensembl Gene IDs, Entrez Gene IDs, etc.

2.  A column containing names. Gene Symbols are commonly used for this purpose.

3.  Columns containing protein abundances of the different samples in the experiment. These columns should have a common prefix (e.g., *abundance.*) used for their identification within the table. Furthermore, replicates are indicated by an underscore followed by a number. For example, *abundance.KT2440_Glc_1*, *abundance.KT2440_Glc_2*, *abundance.KT2440_Glc_3*, and *abundance.KT2440_Glc_4* indicate four replicates for the condition *KT2440_Glc*.

4. *Optional*: Columns containing *+* and *-* indicating contaminant proteins and decoy database hits that should be filtered out.

# Run a Complete Proteomics Analysis Workflow

## Read Data

The function `prot.read_data` accepts either an R dataframe or a table file with extension '.xlsx', '.xls', '.csv', '.tsv', or '.txt'. It creates a `SummarizedExperiment` object and performs feature filtering based on indicated contaminants or unspecific identification, a defined relative standard deviation threshold, or the presence of missing values. For details, run `help("prot.read_data", "VisomX")`. Here, we filter proteins that are not present in at least 6 out of 9 samples.

```{r include = T, eval = T}
# Load the VisomX package
library(VisomX)

# Read data, incl. filtering
data <- prot.read_data(data = system.file("prot_KT2440_glc_ac_LB.txt", package = "VisomX"),
        pfx = "abundance.",
        id = "Ensembl Gene ID",
        name = "Gene Symbol",
        filt_type = "fraction",
        filt_min = 0.66)
```

## Run Workflow



# References

::: {#refs}
:::
