<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en"><head><meta http-equiv="Content-Type" content="text/html; charset=UTF-8"><meta charset="utf-8"><meta http-equiv="X-UA-Compatible" content="IE=edge"><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no"><title>Package index â€¢ VisomX</title><script src="../deps/jquery-3.6.0/jquery-3.6.0.min.js"></script><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no"><link href="../deps/bootstrap-5.3.1/bootstrap.min.css" rel="stylesheet"><script src="../deps/bootstrap-5.3.1/bootstrap.bundle.min.js"></script><link href="../deps/font-awesome-6.5.2/css/all.min.css" rel="stylesheet"><link href="../deps/font-awesome-6.5.2/css/v4-shims.min.css" rel="stylesheet"><script src="../deps/headroom-0.11.0/headroom.min.js"></script><script src="../deps/headroom-0.11.0/jQuery.headroom.min.js"></script><script src="../deps/bootstrap-toc-1.0.1/bootstrap-toc.min.js"></script><script src="../deps/clipboard.js-2.0.11/clipboard.min.js"></script><script src="../deps/search-1.0.0/autocomplete.jquery.min.js"></script><script src="../deps/search-1.0.0/fuse.min.js"></script><script src="../deps/search-1.0.0/mark.min.js"></script><!-- pkgdown --><script src="../pkgdown.js"></script><link href="../extra.css" rel="stylesheet"><meta property="og:title" content="Package index"></head><body>
    <a href="#main" class="visually-hidden-focusable">Skip to contents</a>


    <nav class="navbar navbar-expand-lg fixed-top bg-light" data-bs-theme="light" aria-label="Site navigation"><div class="container">

    <a class="navbar-brand me-2" href="../index.html">VisomX</a>

    <small class="nav-text text-muted me-auto" data-bs-toggle="tooltip" data-bs-placement="bottom" title="">0.0.94</small>


    <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbar" aria-controls="navbar" aria-expanded="false" aria-label="Toggle navigation">
      <span class="navbar-toggler-icon"></span>
    </button>

    <div id="navbar" class="collapse navbar-collapse ms-3">
      <ul class="navbar-nav me-auto"><li class="active nav-item"><a class="nav-link" href="../reference/index.html">Reference</a></li>
<li class="nav-item dropdown">
  <button class="nav-link dropdown-toggle" type="button" id="dropdown-articles" data-bs-toggle="dropdown" aria-expanded="false" aria-haspopup="true">Articles</button>
  <ul class="dropdown-menu" aria-labelledby="dropdown-articles"><li><a class="dropdown-item" href="../articles/vignette_proteomics.html">Reliable and extensive analysis of Proteomics data with package `VisomX`</a></li>
    <li><a class="dropdown-item" href="../articles/vignette_transcriptomics.html">Reliable and extensive analysis of transcriptomics data with package `VisomX`</a></li>
  </ul></li>
      </ul><ul class="navbar-nav"><li class="nav-item"><form class="form-inline" role="search">
 <input class="form-control" type="search" name="search-input" id="search-input" autocomplete="off" aria-label="Search site" placeholder="Search for" data-search-index="../search.json"></form></li>
<li class="nav-item"><a class="external-link nav-link" href="https://github.com/NicWir/VisomX/" aria-label="GitHub"><span class="fa fab fa-github fa-lg"></span></a></li>
      </ul></div>


  </div>
</nav><div class="container template-reference-index">
<div class="row">
  <main id="main" class="col-md-9"><div class="page-header">

      <h1>Package index</h1>
    </div>

    <div class="section level2">
      <h2 id="all-functions">All functions<a class="anchor" aria-label="anchor" href="#all-functions"></a></h2>




    </div><div class="section level2">




      <dl><dt>

          <code><a href="Get.bwss.html">Get.bwss()</a></code>

        </dt>
        <dd>Compute within group and between group sum of squares (BSS/WSS) for each row of a matrix which may have NA</dd>
      </dl><dl><dt>

          <code><a href="Intersect.html">Intersect()</a></code>

        </dt>
        <dd>Multiple set version of intersect</dd>
      </dl><dl><dt>

          <code><a href="Perform.permutation.html">Perform.permutation()</a></code>

        </dt>
        <dd>Permutation</dd>
      </dl><dl><dt>

          <code><a href="ReplaceMissingByLoD.html">ReplaceMissingByLoD()</a></code>

        </dt>
        <dd>Replace missing values with lower detection limit</dd>
      </dl><dl><dt>

          <code><a href="Setdiff.html">Setdiff()</a></code>

        </dt>
        <dd>Remove the union of the y's from the common x's.</dd>
      </dl><dl><dt>

          <code><a href="Union.html">Union()</a></code>

        </dt>
        <dd>Multiple set version of union</dd>
      </dl><dl><dt>

          <code><a href="colFmt.html">colFmt()</a></code>

        </dt>
        <dd>Format font color for Markdown reports</dd>
      </dl><dl><dt>

          <code><a href="correlate_features_with_vector.html">correlate_features_with_vector()</a></code>

        </dt>
        <dd>Correlate each row (feature) of a matrix with a numeric vector.</dd>
      </dl><dl><dt>

          <code><a href="correlate_matrix_with_covariates.html">correlate_matrix_with_covariates()</a></code>

        </dt>
        <dd>High-level wrapper to correlate all features with multiple covariates.</dd>
      </dl><dl><dt>

          <code><a href="correlate_pathways.html">correlate_pathways()</a></code>

        </dt>
        <dd>Compute simple pathway scores (mean expression) and correlate with covariates.</dd>
      </dl><dl><dt>

          <code><a href="delete_prefix.html">delete_prefix()</a></code>

        </dt>
        <dd>Delete Prefix</dd>
      </dl><dl><dt>

          <code><a href="dot-onAttach.html">.onAttach()</a></code>

        </dt>
        <dd>Run upon attaching package VisomX</dd>
      </dl><dl><dt>

          <code><a href="enrich_pathways.html">enrich_pathways()</a></code>

        </dt>
        <dd>Enrichment analysis of pathways</dd>
      </dl><dl><dt>

          <code><a href="fast.write.csv.html">fast.write.csv()</a></code>

        </dt>
        <dd>Write object in CSV file</dd>
      </dl><dl><dt>

          <code><a href="filter_missval.html">filter_missval()</a></code>

        </dt>
        <dd>Filter SummarizedExperiment object based on missing values</dd>
      </dl><dl><dt>

          <code><a href="flux_to_map.html">flux_to_map()</a></code>

        </dt>
        <dd>Create a metabolic map with fluxes and export as SVG file</dd>
      </dl><dl><dt>

          <code><a href="get_annotation_contrast.html">get_annotation_contrast()</a></code>

        </dt>
        <dd>Subset and annotate a DESeqDataSet for contrast-specific heatmaps</dd>
      </dl><dl><dt>

          <code><a href="get_prefix.html">get_prefix()</a></code>

        </dt>
        <dd>Get the common prefix of words</dd>
      </dl><dl><dt>

          <code><a href="gsea_custom.html">gsea_custom()</a></code>

        </dt>
        <dd>GSEA Custom</dd>
      </dl><dl><dt>

          <code><a href="gsea_kegg.html">gsea_kegg()</a></code>

        </dt>
        <dd>GSEA KEGG</dd>
      </dl><dl><dt>

          <code><a href="make_se.html">make_se()</a></code>

        </dt>
        <dd>Create a SummarizedExperiment object from a data frame</dd>
      </dl><dl><dt>

          <code><a href="manual_impute.html">manual_impute()</a></code>

        </dt>
        <dd>Imputation by random draws from a manually defined distribution</dd>
      </dl><dl><dt>

          <code><a href="meanSdPlot.html">meanSdPlot()</a></code>

        </dt>
        <dd>Plot row standard deviations versus row means</dd>
      </dl><dl><dt>

          <code><a href="met.ANOVA.Anal.html">met.ANOVA.Anal()</a></code>

        </dt>
        <dd>Perform ANOVA analysis</dd>
      </dl><dl><dt>

          <code><a href="met.FC.Anal.html">met.FC.Anal()</a></code>

        </dt>
        <dd>Fold change analysis, unpaired</dd>
      </dl><dl><dt>

          <code><a href="met.FilterVariable.html">met.FilterVariable()</a></code>

        </dt>
        <dd>Methods for non-specific filtering of variables</dd>
      </dl><dl><dt>

          <code><a href="met.GetFC.html">met.GetFC()</a></code>

        </dt>
        <dd>Used by higher functions to calculate fold change</dd>
      </dl><dl><dt>

          <code><a href="met.GetTtestRes.html">met.GetTtestRes()</a></code>

        </dt>
        <dd>Used by higher functions to retrieve T-test p-values</dd>
      </dl><dl><dt>

          <code><a href="met.PCA.Anal.html">met.PCA.Anal()</a></code>

        </dt>
        <dd>Perform a Principal Component Analysis</dd>
      </dl><dl><dt>

          <code><a href="met.PLSDA.CV.html">met.PLSDA.CV()</a></code>

        </dt>
        <dd>PLS-DA classification and feature selection</dd>
      </dl><dl><dt>

          <code><a href="met.PLSDA.Permut.html">met.PLSDA.Permut()</a></code>

        </dt>
        <dd>Perform PLS-DA permutation</dd>
      </dl><dl><dt>

          <code><a href="met.PLSR.Anal.html">met.PLSR.Anal()</a></code>

        </dt>
        <dd>Partial least squares (PLS) analysis using <em>oscorespls</em> (Orthogonal scores algorithm)</dd>
      </dl><dl><dt>

          <code><a href="met.PerformFeatureFilter.html">met.PerformFeatureFilter()</a></code>

        </dt>
        <dd>Used by higher functions for non-specific filtering of variables</dd>
      </dl><dl><dt>

          <code><a href="met.PreparePrenormData.html">met.PreparePrenormData()</a></code>

        </dt>
        <dd>Prepare data for normalization</dd>
      </dl><dl><dt>

          <code><a href="met.SanityCheck.html">met.SanityCheck()</a></code>

        </dt>
        <dd>Perform sanity check on metabolomics data</dd>
      </dl><dl><dt>

          <code><a href="met.Ttests.Anal.html">met.Ttests.Anal()</a></code>

        </dt>
        <dd>Perform T-test analysis</dd>
      </dl><dl><dt>

          <code><a href="met.UpdateData.html">met.UpdateData()</a></code>

        </dt>
        <dd>Remove defined samples and features from the metabolomics dataset.</dd>
      </dl><dl><dt>

          <code><a href="met.impute.html">met.impute()</a></code>

        </dt>
        <dd>Impute missing variables</dd>
      </dl><dl><dt>

          <code><a href="met.initialize.html">met.initialize()</a></code>

        </dt>
        <dd>Constructs a dataSet object for storing metabolomics data</dd>
      </dl><dl><dt>

          <code><a href="met.normalize.html">met.normalize()</a></code>

        </dt>
        <dd>Data normalization</dd>
      </dl><dl><dt>

          <code><a href="met.plot_ANOVA.html">met.plot_ANOVA()</a></code>

        </dt>
        <dd>Visualize p values determined by ANOVA analysis.</dd>
      </dl><dl><dt>

          <code><a href="met.plot_CorrHeatMap_Features.html">met.plot_CorrHeatMap_Features()</a></code>

        </dt>
        <dd>Pattern hunter, plot correlation heatmap</dd>
      </dl><dl><dt>

          <code><a href="met.plot_CorrHeatMap_Samples.html">met.plot_CorrHeatMap_Samples()</a></code>

        </dt>
        <dd>Pattern hunter, plot correlation heatmap</dd>
      </dl><dl><dt>

          <code><a href="met.plot_FeatureNormSummary.html">met.plot_FeatureNormSummary()</a></code>

        </dt>
        <dd>Visualize effect of normalization on feature density distributions.</dd>
      </dl><dl><dt>

          <code><a href="met.plot_ImpVar.html">met.plot_ImpVar()</a></code>

        </dt>
        <dd>Plot PLS important features</dd>
      </dl><dl><dt>

          <code><a href="met.plot_PCA2DLoading.html">met.plot_PCA2DLoading()</a></code>

        </dt>
        <dd>Plot PCA loadings and also set up the matrix for display</dd>
      </dl><dl><dt>

          <code><a href="met.plot_PCA2DScore.html">met.plot_PCA2DScore()</a></code>

        </dt>
        <dd>Create 2D PCA score plot</dd>
      </dl><dl><dt>

          <code><a href="met.plot_PCA3DLoading.html">met.plot_PCA3DLoading()</a></code>

        </dt>
        <dd>Create 3D PCA loading plot</dd>
      </dl><dl><dt>

          <code><a href="met.plot_PCA3DScore.html">met.plot_PCA3DScore()</a></code>

        </dt>
        <dd>Create 3D PCA Score plot</dd>
      </dl><dl><dt>

          <code><a href="met.plot_PCAScree.html">met.plot_PCAScree()</a></code>

        </dt>
        <dd>Plot PCA scree plot</dd>
      </dl><dl><dt>

          <code><a href="met.plot_PLS.Crossvalidation.html">met.plot_PLS.Crossvalidation()</a></code>

        </dt>
        <dd>Plot PLS-DA classification performance using different components</dd>
      </dl><dl><dt>

          <code><a href="met.plot_PLS.Permutation.html">met.plot_PLS.Permutation()</a></code>

        </dt>
        <dd>Plot PLS-DA classification performance using different components, permutation</dd>
      </dl><dl><dt>

          <code><a href="met.plot_PLS2DLoading.html">met.plot_PLS2DLoading()</a></code>

        </dt>
        <dd>Plot PLS-DA loadings and also set up the matrix for display</dd>
      </dl><dl><dt>

          <code><a href="met.plot_PLS2DScore.html">met.plot_PLS2DScore()</a></code>

        </dt>
        <dd>Create 2D PLS-DA score plot</dd>
      </dl><dl><dt>

          <code><a href="met.plot_PLS3DLoading.html">met.plot_PLS3DLoading()</a></code>

        </dt>
        <dd>Create 3D PLS-DA loading plot</dd>
      </dl><dl><dt>

          <code><a href="met.plot_PLS3DScore.html">met.plot_PLS3DScore()</a></code>

        </dt>
        <dd>Create 3D PLS-DA Score plot</dd>
      </dl><dl><dt>

          <code><a href="met.plot_PLSImpScatter.html">met.plot_PLSImpScatter()</a></code>

        </dt>
        <dd>Plot PLS important features</dd>
      </dl><dl><dt>

          <code><a href="met.plot_PLS_Imp.html">met.plot_PLS_Imp()</a></code>

        </dt>
        <dd>Plot PLS important features</dd>
      </dl><dl><dt>

          <code><a href="met.plot_SampleNormSummary.html">met.plot_SampleNormSummary()</a></code>

        </dt>
        <dd>Visualize effect of normalization on sample density distributions.</dd>
      </dl><dl><dt>

          <code><a href="met.plot_detect.html">met.plot_detect()</a></code>

        </dt>
        <dd>Visualize intensities of compounds with missing values</dd>
      </dl><dl><dt>

          <code><a href="met.plot_heatmap.html">met.plot_heatmap()</a></code>

        </dt>
        <dd>Create heat map of hierarchically clustered features and samples</dd>
      </dl><dl><dt>

          <code><a href="met.plot_missval.html">met.plot_missval()</a></code>

        </dt>
        <dd>Plot a heatmap of compounds with missing values</dd>
      </dl><dl><dt>

          <code><a href="met.plot_volcano.html">met.plot_volcano()</a></code>

        </dt>
        <dd>Volcano Plot</dd>
      </dl><dl><dt>

          <code><a href="met.print_PCA3DLoading.html">met.print_PCA3DLoading()</a></code>

        </dt>
        <dd>Plot a generated 3D PCA loading plot in RStudio</dd>
      </dl><dl><dt>

          <code><a href="met.print_PCA3DScore.html">met.print_PCA3DScore()</a></code>

        </dt>
        <dd>Plot a generated 3D PCA scores plot in RStudio</dd>
      </dl><dl><dt>

          <code><a href="met.print_PLS3DLoading.html">met.print_PLS3DLoading()</a></code>

        </dt>
        <dd>Plot a generated 3D PLS-DA loading plot in RStudio</dd>
      </dl><dl><dt>

          <code><a href="met.print_PLS3DScore.html">met.print_PLS3DScore()</a></code>

        </dt>
        <dd>Plot a generated 3D PLS-DA scores plot in RStudio</dd>
      </dl><dl><dt>

          <code><a href="met.read_data.html">met.read_data()</a></code>

        </dt>
        <dd>Construct mSet data container, read metabolomics data, filter data, and impute missing values</dd>
      </dl><dl><dt>

          <code><a href="met.report.html">met.report()</a></code>

        </dt>
        <dd>Generate a markdown report for metabolomics data analysis</dd>
      </dl><dl><dt>

          <code><a href="met.report_test_normalization.html">met.report_test_normalization()</a></code>

        </dt>
        <dd>Generate a markdown report for the screening of metabolomics data pre-processing methods.</dd>
      </dl><dl><dt>

          <code><a href="met.test_normalization.html">met.test_normalization()</a></code>

        </dt>
        <dd>Test for optimal data processing conditions</dd>
      </dl><dl><dt>

          <code><a href="met.workflow.html">met.workflow()</a></code>

        </dt>
        <dd>Perform metabolomics data analysis workflow</dd>
      </dl><dl><dt>

          <code><a href="met_to_map.html">met_to_map()</a></code>

        </dt>
        <dd>Create a Metabolic Map from Metabolite Concentrations</dd>
      </dl><dl><dt>

          <code><a href="pathway_enrich.html">pathway_enrich()</a></code>

        </dt>
        <dd>Pathway Enrichment Analysis</dd>
      </dl><dl><dt>

          <code><a href="plot_coverage.html">plot_coverage()</a></code>

        </dt>
        <dd>Plot coverage of a SummarizedExperiment</dd>
      </dl><dl><dt>

          <code><a href="png_to_gif.html">png_to_gif()</a></code>

        </dt>
        <dd>Convert PNGs to an animated GIF</dd>
      </dl><dl><dt>

          <code><a href="prot.add_rejections.html">prot.add_rejections()</a></code>

        </dt>
        <dd>Add rejections to a SummarizedExperiment</dd>
      </dl><dl><dt>

          <code><a href="prot.boxplot_intensity.html">prot.boxplot_intensity()</a></code>

        </dt>
        <dd>Boxplot Intensity</dd>
      </dl><dl><dt>

          <code><a href="prot.filter_missing.html">prot.filter_missing()</a></code>

        </dt>
        <dd>Filter proteins by missing value criteria</dd>
      </dl><dl><dt>

          <code><a href="prot.get_kegg_pathways.html">prot.get_kegg_pathways()</a></code>

        </dt>
        <dd>Multiple set version of union</dd>
      </dl><dl><dt>

          <code><a href="prot.get_pathway_genes.html">prot.get_pathway_genes()</a></code>

        </dt>
        <dd>Get genes from pathway table</dd>
      </dl><dl><dt>

          <code><a href="prot.get_results.html">prot.get_results()</a></code>

        </dt>
        <dd>Get Results</dd>
      </dl><dl><dt>

          <code><a href="prot.impute.html">prot.impute()</a></code>

        </dt>
        <dd>Impute missing values in a SummarizedExperiment object</dd>
      </dl><dl><dt>

          <code><a href="prot.make_unique.html">prot.make_unique()</a></code>

        </dt>
        <dd>Make Unique Proteins</dd>
      </dl><dl><dt>

          <code><a href="prot.normalize_vsn.html">prot.normalize_vsn()</a></code>

        </dt>
        <dd>Normalize the data via variance stabilization normalization</dd>
      </dl><dl><dt>

          <code><a href="prot.pca.html">prot.pca()</a></code>

        </dt>
        <dd>PCA Analysis</dd>
      </dl><dl><dt>

          <code><a href="prot.plot_bar.html">prot.plot_bar()</a></code>

        </dt>
        <dd>Plot bar plots for protein abundance or fold change</dd>
      </dl><dl><dt>

          <code><a href="prot.plot_corrheatmap.html">prot.plot_corrheatmap()</a></code>

        </dt>
        <dd>Plot a correlation heatmap</dd>
      </dl><dl><dt>

          <code><a href="prot.plot_density.html">prot.plot_density()</a></code>

        </dt>
        <dd>Generates density plots for a SummarizedExperiment object</dd>
      </dl><dl><dt>

          <code><a href="prot.plot_detect.html">prot.plot_detect()</a></code>

        </dt>
        <dd>Plot detectability for a SummarizedExperiment</dd>
      </dl><dl><dt>

          <code><a href="prot.plot_enrichment.html">prot.plot_enrichment()</a></code>

        </dt>
        <dd>Plot pathway enrichment</dd>
      </dl><dl><dt>

          <code><a href="prot.plot_heatmap.html">prot.plot_heatmap()</a></code>

        </dt>
        <dd>Plots a heatmap of the differentially expressed proteins</dd>
      </dl><dl><dt>

          <code><a href="prot.plot_heatmap_all.html">prot.plot_heatmap_all()</a></code>

        </dt>
        <dd>Plots a heatmap of all protein abundances</dd>
      </dl><dl><dt>

          <code><a href="prot.plot_imputation.html">prot.plot_imputation()</a></code>

        </dt>
        <dd>Plot imputation results</dd>
      </dl><dl><dt>

          <code><a href="prot.plot_loadings.html">prot.plot_loadings()</a></code>

        </dt>
        <dd>Plot the loadings of a principal components analysis</dd>
      </dl><dl><dt>

          <code><a href="prot.plot_missval.html">prot.plot_missval()</a></code>

        </dt>
        <dd>Plot missing values in SummarizedExperiment</dd>
      </dl><dl><dt>

          <code><a href="prot.plot_numbers.html">prot.plot_numbers()</a></code>

        </dt>
        <dd>Plot the number of proteins identified in a SummarizedExperiment object</dd>
      </dl><dl><dt>

          <code><a href="prot.plot_pca.html">prot.plot_pca()</a></code>

        </dt>
        <dd>Plot Results of Principal Component Analysis</dd>
      </dl><dl><dt>

          <code><a href="prot.plot_screeplot.html">prot.plot_screeplot()</a></code>

        </dt>
        <dd>SCREE plot</dd>
      </dl><dl><dt>

          <code><a href="prot.plot_upset.html">prot.plot_upset()</a></code>

        </dt>
        <dd>Plot an UpSet enrichment plot</dd>
      </dl><dl><dt>

          <code><a href="prot.plot_volcano.html">prot.plot_volcano()</a></code>

        </dt>
        <dd>Plots a volcano plot for a given contrast</dd>
      </dl><dl><dt>

          <code><a href="prot.read_data.html">prot.read_data()</a></code>

        </dt>
        <dd>Read proteomics data in table format and create SummarizedExperiment</dd>
      </dl><dl><dt>

          <code><a href="prot.report.html">prot.report()</a></code>

        </dt>
        <dd>Report Proteomics Data</dd>
      </dl><dl><dt>

          <code><a href="prot.test_diff.html">prot.test_diff()</a></code>

        </dt>
        <dd>Test Differential Expression</dd>
      </dl><dl><dt>

          <code><a href="prot.workflow.html">prot.workflow()</a></code>

        </dt>
        <dd>Run a complete proteomics analysis workflow.</dd>
      </dl><dl><dt>

          <code><a href="prot_to_map.html">prot_to_map()</a></code>

        </dt>
        <dd>Title</dd>
      </dl><dl><dt>

          <code><a href="read_file.html">read_file()</a></code>

        </dt>
        <dd>Call the appropriate function required to read a table file and return the table as a dataframe object.</dd>
      </dl><dl><dt>

          <code><a href="read_flux.html">read_flux()</a></code>

        </dt>
        <dd>Reads flux data from a file or data frame</dd>
      </dl><dl><dt>

          <code><a href="rna.filter_missing.html">rna.filter_missing()</a></code>

        </dt>
        <dd>Filter genes based on missing values</dd>
      </dl><dl><dt>

          <code><a href="rna.impute.html">rna.impute()</a></code>

        </dt>
        <dd>Impute missing values in a SummarizedExperiment object</dd>
      </dl><dl><dt>

          <code><a href="rna.make_se.html">rna.make_se()</a></code>

        </dt>
        <dd>Create a SummarizedExperiment object</dd>
      </dl><dl><dt>

          <code><a href="rna.plot_bar.html">rna.plot_bar()</a></code>

        </dt>
        <dd>Plot bar plots for gene abundance or fold change</dd>
      </dl><dl><dt>

          <code><a href="rna.plot_corrheatmap.html">rna.plot_corrheatmap()</a></code>

        </dt>
        <dd>Plot a correlation heatmap</dd>
      </dl><dl><dt>

          <code><a href="rna.plot_heatmap.html">rna.plot_heatmap()</a></code>

        </dt>
        <dd>Plots a heatmap of the differentially expressed genes</dd>
      </dl><dl><dt>

          <code><a href="rna.plot_imputation.html">rna.plot_imputation()</a></code>

        </dt>
        <dd>Plot imputation results</dd>
      </dl><dl><dt>

          <code><a href="rna.plot_numbers.html">rna.plot_numbers()</a></code>

        </dt>
        <dd>Plot the number of genes identified in a DESeqDataSet object</dd>
      </dl><dl><dt>

          <code><a href="rna.plot_pca.html">rna.plot_pca()</a></code>

        </dt>
        <dd>Plot Results of Principal Component Analysis</dd>
      </dl><dl><dt>

          <code><a href="rna.plot_volcano.html">rna.plot_volcano()</a></code>

        </dt>
        <dd>Plots a volcano plot for a given contrast</dd>
      </dl><dl><dt>

          <code><a href="rna.read_data.html">rna.read_data()</a></code>

        </dt>
        <dd>Read transcriptomics data</dd>
      </dl><dl><dt>

          <code><a href="rna.report.html">rna.report()</a></code>

        </dt>
        <dd>Generate a transcriptomics analysis report</dd>
      </dl><dl><dt>

          <code><a href="rna.workflow.html">rna.workflow()</a></code>

        </dt>
        <dd>RNA sequencing workflow</dd>
      </dl><dl><dt>

          <code><a href="svg_to_pdf.html">svg_to_pdf()</a></code>

        </dt>
        <dd>Convert SVG to PDF</dd>
      </dl><dl><dt>

          <code><a href="theme_DEP1.html">theme_DEP1()</a></code>

        </dt>
        <dd>theme_DEP1</dd>
      </dl><dl><dt>

          <code><a href="theme_DEP2.html">theme_DEP2()</a></code>

        </dt>
        <dd>Theme DEP2</dd>
      </dl></div>
  </main></div>


    <footer><div class="pkgdown-footer-left">
  <p>Developed by Nicolas T. Wirth.</p>
</div>

<div class="pkgdown-footer-right">
  <p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.1.3.</p>
</div>

    </footer></div>





  </body></html>

