<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en"><head><meta http-equiv="Content-Type" content="text/html; charset=UTF-8"><meta charset="utf-8"><meta http-equiv="X-UA-Compatible" content="IE=edge"><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no"><title>Pathway Enrichment Analysis — pathway_enrich • VisomX</title><script src="../deps/jquery-3.6.0/jquery-3.6.0.min.js"></script><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no"><link href="../deps/bootstrap-5.3.1/bootstrap.min.css" rel="stylesheet"><script src="../deps/bootstrap-5.3.1/bootstrap.bundle.min.js"></script><link href="../deps/font-awesome-6.5.2/css/all.min.css" rel="stylesheet"><link href="../deps/font-awesome-6.5.2/css/v4-shims.min.css" rel="stylesheet"><script src="../deps/headroom-0.11.0/headroom.min.js"></script><script src="../deps/headroom-0.11.0/jQuery.headroom.min.js"></script><script src="../deps/bootstrap-toc-1.0.1/bootstrap-toc.min.js"></script><script src="../deps/clipboard.js-2.0.11/clipboard.min.js"></script><script src="../deps/search-1.0.0/autocomplete.jquery.min.js"></script><script src="../deps/search-1.0.0/fuse.min.js"></script><script src="../deps/search-1.0.0/mark.min.js"></script><!-- pkgdown --><script src="../pkgdown.js"></script><link href="../extra.css" rel="stylesheet"><meta property="og:title" content="Pathway Enrichment Analysis — pathway_enrich"><meta name="description" content="This function performs pathway enrichment analysis. Uses the DOSE package with the internal function enricher_internal."><meta property="og:description" content="This function performs pathway enrichment analysis. Uses the DOSE package with the internal function enricher_internal."></head><body>
    <a href="#main" class="visually-hidden-focusable">Skip to contents</a>


    <nav class="navbar navbar-expand-lg fixed-top bg-light" data-bs-theme="light" aria-label="Site navigation"><div class="container">

    <a class="navbar-brand me-2" href="../index.html">VisomX</a>

    <small class="nav-text text-muted me-auto" data-bs-toggle="tooltip" data-bs-placement="bottom" title="">0.0.93</small>


    <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbar" aria-controls="navbar" aria-expanded="false" aria-label="Toggle navigation">
      <span class="navbar-toggler-icon"></span>
    </button>

    <div id="navbar" class="collapse navbar-collapse ms-3">
      <ul class="navbar-nav me-auto"><li class="active nav-item"><a class="nav-link" href="../reference/index.html">Reference</a></li>
<li class="nav-item dropdown">
  <button class="nav-link dropdown-toggle" type="button" id="dropdown-articles" data-bs-toggle="dropdown" aria-expanded="false" aria-haspopup="true">Articles</button>
  <ul class="dropdown-menu" aria-labelledby="dropdown-articles"><li><a class="dropdown-item" href="../articles/vignette_proteomics.html">Reliable and extensive analysis of Proteomics data with package `VisomX`</a></li>
    <li><a class="dropdown-item" href="../articles/vignette_transcriptomics.html">Reliable and extensive analysis of transcriptomics data with package `VisomX`</a></li>
  </ul></li>
      </ul><ul class="navbar-nav"><li class="nav-item"><form class="form-inline" role="search">
 <input class="form-control" type="search" name="search-input" id="search-input" autocomplete="off" aria-label="Search site" placeholder="Search for" data-search-index="../search.json"></form></li>
<li class="nav-item"><a class="external-link nav-link" href="https://github.com/NicWir/VisomX/" aria-label="GitHub"><span class="fa fab fa-github fa-lg"></span></a></li>
      </ul></div>


  </div>
</nav><div class="container template-reference-topic">
<div class="row">
  <main id="main" class="col-md-9"><div class="page-header">

      <h1>Pathway Enrichment Analysis</h1>
      <small class="dont-index">Source: <a href="https://github.com/NicWir/VisomX/blob/master/R/proteomics_computation.R" class="external-link"><code>R/proteomics_computation.R</code></a></small>
      <div class="d-none name"><code>pathway_enrich.Rd</code></div>
    </div>

    <div class="ref-description section level2">
    <p>This function performs pathway enrichment analysis. Uses the DOSE package with the internal function <code>enricher_internal</code>.</p>
    </div>

    <div class="section level2">
    <h2 id="ref-usage">Usage<a class="anchor" aria-label="anchor" href="#ref-usage"></a></h2>
    <div class="sourceCode"><pre class="sourceCode r"><code><span><span class="fu">pathway_enrich</span><span class="op">(</span></span>
<span>  <span class="va">gene</span>,</span>
<span>  organism <span class="op">=</span> <span class="st">"ppu"</span>,</span>
<span>  keyType <span class="op">=</span> <span class="st">"kegg"</span>,</span>
<span>  pvalueCutoff <span class="op">=</span> <span class="fl">0.05</span>,</span>
<span>  pAdjustMethod <span class="op">=</span> <span class="st">"BH"</span>,</span>
<span>  <span class="va">universe</span>,</span>
<span>  minGSSize <span class="op">=</span> <span class="fl">10</span>,</span>
<span>  maxGSSize <span class="op">=</span> <span class="fl">500</span>,</span>
<span>  qvalueCutoff <span class="op">=</span> <span class="fl">0.2</span>,</span>
<span>  use_internal_kegg <span class="op">=</span> <span class="cn">FALSE</span>,</span>
<span>  custom_gene_sets <span class="op">=</span> <span class="cn">FALSE</span>,</span>
<span>  custom_pathways <span class="op">=</span> <span class="cn">NULL</span></span>
<span><span class="op">)</span></span></code></pre></div>
    </div>

    <div class="section level2">
    <h2 id="arguments">Arguments<a class="anchor" aria-label="anchor" href="#arguments"></a></h2>


<dl><dt id="arg-gene">gene<a class="anchor" aria-label="anchor" href="#arg-gene"></a></dt>
<dd><p>A numeric vector or a character vector of gene symbols that were found enriched in a differential expression analysis.</p></dd>


<dt id="arg-organism">organism<a class="anchor" aria-label="anchor" href="#arg-organism"></a></dt>
<dd><p>A character string. The organism to use. Default is "ppu" for Pseudomonas putida KT2440.</p></dd>


<dt id="arg-keytype">keyType<a class="anchor" aria-label="anchor" href="#arg-keytype"></a></dt>
<dd><p>A character string. The type of gene ID. Default is "kegg".</p></dd>


<dt id="arg-pvaluecutoff">pvalueCutoff<a class="anchor" aria-label="anchor" href="#arg-pvaluecutoff"></a></dt>
<dd><p>A numeric value. The cutoff of p-value. Default is 0.05.</p></dd>


<dt id="arg-padjustmethod">pAdjustMethod<a class="anchor" aria-label="anchor" href="#arg-padjustmethod"></a></dt>
<dd><p>A character string. The method of p-value adjustment. Default is "BH".</p></dd>


<dt id="arg-universe">universe<a class="anchor" aria-label="anchor" href="#arg-universe"></a></dt>
<dd><p>A numeric vector or a character vector of gene symbols.</p></dd>


<dt id="arg-mingssize">minGSSize<a class="anchor" aria-label="anchor" href="#arg-mingssize"></a></dt>
<dd><p>A numeric value. The minimum gene set size. Default is 10.</p></dd>


<dt id="arg-maxgssize">maxGSSize<a class="anchor" aria-label="anchor" href="#arg-maxgssize"></a></dt>
<dd><p>A numeric value. The maximum gene set size. Default is 500.</p></dd>


<dt id="arg-qvaluecutoff">qvalueCutoff<a class="anchor" aria-label="anchor" href="#arg-qvaluecutoff"></a></dt>
<dd><p>A numeric value. The cutoff of q-value. Default is 0.2.</p></dd>


<dt id="arg-use-internal-kegg">use_internal_kegg<a class="anchor" aria-label="anchor" href="#arg-use-internal-kegg"></a></dt>
<dd><p>A logical value. Whether to use internal KEGG database. Default is FALSE.</p></dd>


<dt id="arg-custom-gene-sets">custom_gene_sets<a class="anchor" aria-label="anchor" href="#arg-custom-gene-sets"></a></dt>
<dd><p>A logical value. Whether to use custom gene sets. Default is FALSE.</p></dd>


<dt id="arg-custom-pathways">custom_pathways<a class="anchor" aria-label="anchor" href="#arg-custom-pathways"></a></dt>
<dd><p>A data frame. The custom pathways to use. Needs to have column "Pathway" and corresponding "Accession" column with comma-separated entries of gene/protein identifiers for that pathway.</p></dd>

</dl></div>
    <div class="section level2">
    <h2 id="value">Value<a class="anchor" aria-label="anchor" href="#value"></a></h2>
    <p>A data frame containing the results of the pathway enrichment analysis.</p>
    </div>

  </main><aside class="col-md-3"><nav id="toc" aria-label="Table of contents"><h2>On this page</h2>
    </nav></aside></div>


    <footer><div class="pkgdown-footer-left">
  <p>Developed by Nicolas T. Wirth.</p>
</div>

<div class="pkgdown-footer-right">
  <p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.1.1.</p>
</div>

    </footer></div>





  </body></html>

