<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en"><head><meta http-equiv="Content-Type" content="text/html; charset=UTF-8"><meta charset="utf-8"><meta http-equiv="X-UA-Compatible" content="IE=edge"><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no"><title>Read proteomics data in table format and create SummarizedExperiment — prot.read_data • VisomX</title><script src="../deps/jquery-3.6.0/jquery-3.6.0.min.js"></script><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no"><link href="../deps/bootstrap-5.3.1/bootstrap.min.css" rel="stylesheet"><script src="../deps/bootstrap-5.3.1/bootstrap.bundle.min.js"></script><link href="../deps/font-awesome-6.5.2/css/all.min.css" rel="stylesheet"><link href="../deps/font-awesome-6.5.2/css/v4-shims.min.css" rel="stylesheet"><script src="../deps/headroom-0.11.0/headroom.min.js"></script><script src="../deps/headroom-0.11.0/jQuery.headroom.min.js"></script><script src="../deps/bootstrap-toc-1.0.1/bootstrap-toc.min.js"></script><script src="../deps/clipboard.js-2.0.11/clipboard.min.js"></script><script src="../deps/search-1.0.0/autocomplete.jquery.min.js"></script><script src="../deps/search-1.0.0/fuse.min.js"></script><script src="../deps/search-1.0.0/mark.min.js"></script><!-- pkgdown --><script src="../pkgdown.js"></script><link href="../extra.css" rel="stylesheet"><meta property="og:title" content="Read proteomics data in table format and create SummarizedExperiment — prot.read_data"><meta name="description" content="prot.read_data takes a table file containing proteomics data and filters proteins based on missing values (see prot.filter_missing) or a given relative standard deviation threshold, and creates a SummarizedExperiment object."><meta property="og:description" content="prot.read_data takes a table file containing proteomics data and filters proteins based on missing values (see prot.filter_missing) or a given relative standard deviation threshold, and creates a SummarizedExperiment object."></head><body>
    <a href="#main" class="visually-hidden-focusable">Skip to contents</a>


    <nav class="navbar navbar-expand-lg fixed-top bg-light" data-bs-theme="light" aria-label="Site navigation"><div class="container">

    <a class="navbar-brand me-2" href="../index.html">VisomX</a>

    <small class="nav-text text-muted me-auto" data-bs-toggle="tooltip" data-bs-placement="bottom" title="">0.0.94</small>


    <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbar" aria-controls="navbar" aria-expanded="false" aria-label="Toggle navigation">
      <span class="navbar-toggler-icon"></span>
    </button>

    <div id="navbar" class="collapse navbar-collapse ms-3">
      <ul class="navbar-nav me-auto"><li class="active nav-item"><a class="nav-link" href="../reference/index.html">Reference</a></li>
<li class="nav-item dropdown">
  <button class="nav-link dropdown-toggle" type="button" id="dropdown-articles" data-bs-toggle="dropdown" aria-expanded="false" aria-haspopup="true">Articles</button>
  <ul class="dropdown-menu" aria-labelledby="dropdown-articles"><li><a class="dropdown-item" href="../articles/vignette_proteomics.html">Reliable and extensive analysis of Proteomics data with package `VisomX`</a></li>
    <li><a class="dropdown-item" href="../articles/vignette_transcriptomics.html">Reliable and extensive analysis of transcriptomics data with package `VisomX`</a></li>
  </ul></li>
      </ul><ul class="navbar-nav"><li class="nav-item"><form class="form-inline" role="search">
 <input class="form-control" type="search" name="search-input" id="search-input" autocomplete="off" aria-label="Search site" placeholder="Search for" data-search-index="../search.json"></form></li>
<li class="nav-item"><a class="external-link nav-link" href="https://github.com/NicWir/VisomX/" aria-label="GitHub"><span class="fa fab fa-github fa-lg"></span></a></li>
      </ul></div>


  </div>
</nav><div class="container template-reference-topic">
<div class="row">
  <main id="main" class="col-md-9"><div class="page-header">

      <h1>Read proteomics data in table format and create SummarizedExperiment</h1>
      <small class="dont-index">Source: <a href="https://github.com/NicWir/VisomX/blob/master/R/proteomics_computation.R" class="external-link"><code>R/proteomics_computation.R</code></a></small>
      <div class="d-none name"><code>prot.read_data.Rd</code></div>
    </div>

    <div class="ref-description section level2">
    <p><code>prot.read_data</code> takes a table file containing proteomics data and filters proteins based on missing values (see <code><a href="prot.filter_missing.html">prot.filter_missing</a></code>) or a given relative standard deviation threshold, and creates a <code>SummarizedExperiment</code> object.</p>
    </div>

    <div class="section level2">
    <h2 id="ref-usage">Usage<a class="anchor" aria-label="anchor" href="#ref-usage"></a></h2>
    <div class="sourceCode"><pre class="sourceCode r"><code><span><span class="fu">prot.read_data</span><span class="op">(</span></span>
<span>  data <span class="op">=</span> <span class="st">"dat_prot.csv"</span>,</span>
<span>  expdesign <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  csvsep <span class="op">=</span> <span class="st">";"</span>,</span>
<span>  dec <span class="op">=</span> <span class="st">"."</span>,</span>
<span>  na.strings <span class="op">=</span> <span class="st">""</span>,</span>
<span>  sheet <span class="op">=</span> <span class="fl">1</span>,</span>
<span>  filter <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"Reverse"</span>, <span class="st">"Potential contaminant"</span><span class="op">)</span>,</span>
<span>  rsd_thresh <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  name <span class="op">=</span> <span class="st">"Gene Symbol"</span>,</span>
<span>  id <span class="op">=</span> <span class="st">"Ensembl Gene ID"</span>,</span>
<span>  pfx <span class="op">=</span> <span class="st">"abundances."</span>,</span>
<span>  filt_type <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"condition"</span>, <span class="st">"complete"</span>, <span class="st">"fraction"</span>, <span class="cn">NULL</span><span class="op">)</span>,</span>
<span>  filt_thr <span class="op">=</span> <span class="fl">3</span>,</span>
<span>  filt_min <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  log2_transform <span class="op">=</span> <span class="cn">TRUE</span></span>
<span><span class="op">)</span></span></code></pre></div>
    </div>

    <div class="section level2">
    <h2 id="arguments">Arguments<a class="anchor" aria-label="anchor" href="#arguments"></a></h2>


<dl><dt id="arg-data">data<a class="anchor" aria-label="anchor" href="#arg-data"></a></dt>
<dd><p>An R dataframe object or a table file with extension '.xlsx', '.xls', '.csv', '.tsv', or '.txt' containing proteomics data.
The table must contain:</p><ol><li><p>A column with protein IDs (e.g., accession numbers). The header of this column is provided as argument <code>id</code>.</p></li>
<li><p>A column with protein names. The header of this column is provided as argument <code>name</code>.</p></li>
<li><p>X columns containing abundance values for X samples. The column headers must have a prefix (e.g., "abundances.") that is provided as argument <code>pfx</code>. Replicates are identified by identical column headers followed by an underscore and the replicate number (e.g., "abundances.ConditionA_1", "abundances.ConditionA_2", "abundances.ConditionA_3", ...).</p></li>
</ol></dd>


<dt id="arg-expdesign">expdesign<a class="anchor" aria-label="anchor" href="#arg-expdesign"></a></dt>
<dd><p>(<em>optional, if made previously</em>) An R dataframe object or a table file containing the columns 'label', 'condition', and 'replicate' with label = "condition_replicate". If <code>NULL</code>, an experimental design table will be created automatically.</p></dd>


<dt id="arg-csvsep">csvsep<a class="anchor" aria-label="anchor" href="#arg-csvsep"></a></dt>
<dd><p>(Character string) separator used in CSV files (ignored for other file types). Default: <code>";"</code></p></dd>


<dt id="arg-dec">dec<a class="anchor" aria-label="anchor" href="#arg-dec"></a></dt>
<dd><p>(Character string) decimal separator used in CSV, TSV or TXT files (ignored for other file types). Default: <code>"."</code></p></dd>


<dt id="arg-na-strings">na.strings<a class="anchor" aria-label="anchor" href="#arg-na-strings"></a></dt>
<dd><p>A character vector of strings which are to be interpreted as NA values.</p></dd>


<dt id="arg-sheet">sheet<a class="anchor" aria-label="anchor" href="#arg-sheet"></a></dt>
<dd><p>(Integer or Character string) Number or name of the sheet with proteomics data in XLS or XLSX files (<em>optional</em>).</p></dd>


<dt id="arg-filter">filter<a class="anchor" aria-label="anchor" href="#arg-filter"></a></dt>
<dd><p>(Character string or vector of strings) Provide the header of a column containing "+" or "-" to indicate if proteins should be discarded or kept, respectively.</p></dd>


<dt id="arg-rsd-thresh">rsd_thresh<a class="anchor" aria-label="anchor" href="#arg-rsd-thresh"></a></dt>
<dd><p>(Numeric, optional) Provide a relative standard deviation (RSD) threshold <strong>in %</strong> for proteins. The RSD is calculated for each condition and if the maximum RSD value determined for a given protein exceeds <code>rsd_thresh</code>, the protein is discarded. The RSD filter is applied <strong>before</strong> further missing value filters based on the three <code>filt_</code> arguments.</p></dd>


<dt id="arg-name">name<a class="anchor" aria-label="anchor" href="#arg-name"></a></dt>
<dd><p>(Character string) Provide the header of the column containing protein names.</p></dd>


<dt id="arg-id">id<a class="anchor" aria-label="anchor" href="#arg-id"></a></dt>
<dd><p>(Character string) Provide the header of the column containing protein IDs</p></dd>


<dt id="arg-pfx">pfx<a class="anchor" aria-label="anchor" href="#arg-pfx"></a></dt>
<dd><p>(Character string) Provide the common prefix for headers containing abundance values (e.g., "abundances.").</p></dd>


<dt id="arg-filt-type">filt_type<a class="anchor" aria-label="anchor" href="#arg-filt-type"></a></dt>
<dd><p>(Character string) "complete", "condition" or "fraction", Sets the type of filtering applied. "complete" will only keep proteins with valid values in all samples. "condition" will keep proteins that have a maximum of <code>filt_thr</code> missing values in at least one condition. "fraction" will keep proteins that have a <code>filt_min</code> fraction of valid values in all samples.</p></dd>


<dt id="arg-filt-thr">filt_thr<a class="anchor" aria-label="anchor" href="#arg-filt-thr"></a></dt>
<dd><p>(Integer) Sets the threshold for the allowed number of missing values in at least one condition if <code>filt_type = "condition"</code>. In other words: "keep proteins that have a maximum of 'filt_thr' missing values in at least one condition."</p></dd>


<dt id="arg-filt-min">filt_min<a class="anchor" aria-label="anchor" href="#arg-filt-min"></a></dt>
<dd><p>(Numeric) Sets the threshold for the minimum fraction of valid values allowed for any protein if <code>filt_type = "fraction"</code>.</p></dd>


<dt id="arg-log-transform">log2_transform<a class="anchor" aria-label="anchor" href="#arg-log-transform"></a></dt>
<dd><p>(Logical) Should the data be log2 transformed?</p></dd>

</dl></div>
    <div class="section level2">
    <h2 id="value">Value<a class="anchor" aria-label="anchor" href="#value"></a></h2>
    <p>A filtered SummarizedExperiment object.</p>
    </div>

  </main><aside class="col-md-3"><nav id="toc" aria-label="Table of contents"><h2>On this page</h2>
    </nav></aside></div>


    <footer><div class="pkgdown-footer-left">
  <p>Developed by Nicolas T. Wirth.</p>
</div>

<div class="pkgdown-footer-right">
  <p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.1.3.</p>
</div>

    </footer></div>





  </body></html>

