<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en"><head><meta http-equiv="Content-Type" content="text/html; charset=UTF-8"><meta charset="utf-8"><meta http-equiv="X-UA-Compatible" content="IE=edge"><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no"><title>Run a complete proteomics analysis workflow. — prot.workflow • VisomX</title><script src="../deps/jquery-3.6.0/jquery-3.6.0.min.js"></script><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no"><link href="../deps/bootstrap-5.3.1/bootstrap.min.css" rel="stylesheet"><script src="../deps/bootstrap-5.3.1/bootstrap.bundle.min.js"></script><link href="../deps/font-awesome-6.5.2/css/all.min.css" rel="stylesheet"><link href="../deps/font-awesome-6.5.2/css/v4-shims.min.css" rel="stylesheet"><script src="../deps/headroom-0.11.0/headroom.min.js"></script><script src="../deps/headroom-0.11.0/jQuery.headroom.min.js"></script><script src="../deps/bootstrap-toc-1.0.1/bootstrap-toc.min.js"></script><script src="../deps/clipboard.js-2.0.11/clipboard.min.js"></script><script src="../deps/search-1.0.0/autocomplete.jquery.min.js"></script><script src="../deps/search-1.0.0/fuse.min.js"></script><script src="../deps/search-1.0.0/mark.min.js"></script><!-- pkgdown --><script src="../pkgdown.js"></script><link href="../extra.css" rel="stylesheet"><meta property="og:title" content="Run a complete proteomics analysis workflow. — prot.workflow"><meta name="description" content="prot.workflow performs variance stabilization normalization (prot.normalize_vsn), missing value imputation (prot.impute), principal component analysis (prot.pca), differential enrichment test (prot.test_diff), and pathway enrichment analysis. If desired, standardized plots and a report are generated and exported as separate files."><meta property="og:description" content="prot.workflow performs variance stabilization normalization (prot.normalize_vsn), missing value imputation (prot.impute), principal component analysis (prot.pca), differential enrichment test (prot.test_diff), and pathway enrichment analysis. If desired, standardized plots and a report are generated and exported as separate files."></head><body>
    <a href="#main" class="visually-hidden-focusable">Skip to contents</a>


    <nav class="navbar navbar-expand-lg fixed-top bg-light" data-bs-theme="light" aria-label="Site navigation"><div class="container">

    <a class="navbar-brand me-2" href="../index.html">VisomX</a>

    <small class="nav-text text-muted me-auto" data-bs-toggle="tooltip" data-bs-placement="bottom" title="">0.0.93</small>


    <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbar" aria-controls="navbar" aria-expanded="false" aria-label="Toggle navigation">
      <span class="navbar-toggler-icon"></span>
    </button>

    <div id="navbar" class="collapse navbar-collapse ms-3">
      <ul class="navbar-nav me-auto"><li class="active nav-item"><a class="nav-link" href="../reference/index.html">Reference</a></li>
<li class="nav-item dropdown">
  <button class="nav-link dropdown-toggle" type="button" id="dropdown-articles" data-bs-toggle="dropdown" aria-expanded="false" aria-haspopup="true">Articles</button>
  <ul class="dropdown-menu" aria-labelledby="dropdown-articles"><li><a class="dropdown-item" href="../articles/vignette_proteomics.html">Reliable and extensive analysis of Proteomics data with package `VisomX`</a></li>
    <li><a class="dropdown-item" href="../articles/vignette_transcriptomics.html">Reliable and extensive analysis of transcriptomics data with package `VisomX`</a></li>
  </ul></li>
      </ul><ul class="navbar-nav"><li class="nav-item"><form class="form-inline" role="search">
 <input class="form-control" type="search" name="search-input" id="search-input" autocomplete="off" aria-label="Search site" placeholder="Search for" data-search-index="../search.json"></form></li>
<li class="nav-item"><a class="external-link nav-link" href="https://github.com/NicWir/VisomX/" aria-label="GitHub"><span class="fa fab fa-github fa-lg"></span></a></li>
      </ul></div>


  </div>
</nav><div class="container template-reference-topic">
<div class="row">
  <main id="main" class="col-md-9"><div class="page-header">

      <h1>Run a complete proteomics analysis workflow.</h1>
      <small class="dont-index">Source: <a href="https://github.com/NicWir/VisomX/blob/master/R/proteomics_computation.R" class="external-link"><code>R/proteomics_computation.R</code></a></small>
      <div class="d-none name"><code>prot.workflow.Rd</code></div>
    </div>

    <div class="ref-description section level2">
    <p><code>prot.workflow</code> performs variance stabilization normalization (<code><a href="prot.normalize_vsn.html">prot.normalize_vsn</a></code>), missing value imputation (<code><a href="prot.impute.html">prot.impute</a></code>), principal component analysis (<code><a href="prot.pca.html">prot.pca</a></code>), differential enrichment test (<code><a href="prot.test_diff.html">prot.test_diff</a></code>), and pathway enrichment analysis. If desired, standardized plots and a report are generated and exported as separate files.</p>
    </div>

    <div class="section level2">
    <h2 id="ref-usage">Usage<a class="anchor" aria-label="anchor" href="#ref-usage"></a></h2>
    <div class="sourceCode"><pre class="sourceCode r"><code><span><span class="fu">prot.workflow</span><span class="op">(</span></span>
<span>  <span class="va">se</span>,</span>
<span>  imp_fun <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"SampMin"</span>, <span class="st">"man"</span>, <span class="st">"bpca"</span>, <span class="st">"knn"</span>, <span class="st">"QRILC"</span>, <span class="st">"MLE"</span>, <span class="st">"MinDet"</span>, <span class="st">"MinProb"</span>,</span>
<span>    <span class="st">"min"</span>, <span class="st">"zero"</span>, <span class="st">"mixed"</span>, <span class="st">"nbavg"</span><span class="op">)</span>,</span>
<span>  q <span class="op">=</span> <span class="fl">0.01</span>,</span>
<span>  knn.rowmax <span class="op">=</span> <span class="fl">0.5</span>,</span>
<span>  type <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"all"</span>, <span class="st">"control"</span>, <span class="st">"manual"</span><span class="op">)</span>,</span>
<span>  control <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  contrast <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  alpha <span class="op">=</span> <span class="fl">0.05</span>,</span>
<span>  alpha_pathways <span class="op">=</span> <span class="fl">0.1</span>,</span>
<span>  lfc <span class="op">=</span> <span class="fl">1</span>,</span>
<span>  heatmap.show_all <span class="op">=</span> <span class="cn">TRUE</span>,</span>
<span>  heatmap.kmeans <span class="op">=</span> <span class="cn">F</span>,</span>
<span>  k <span class="op">=</span> <span class="fl">6</span>,</span>
<span>  heatmap.col_limit <span class="op">=</span> <span class="cn">NA</span>,</span>
<span>  heatmap.show_row_names <span class="op">=</span> <span class="cn">TRUE</span>,</span>
<span>  heatmap.row_font_size <span class="op">=</span> <span class="fl">6</span>,</span>
<span>  volcano.add_names <span class="op">=</span> <span class="cn">FALSE</span>,</span>
<span>  volcano.label_size <span class="op">=</span> <span class="fl">2.5</span>,</span>
<span>  volcano.adjusted <span class="op">=</span> <span class="cn">TRUE</span>,</span>
<span>  plot <span class="op">=</span> <span class="cn">FALSE</span>,</span>
<span>  export <span class="op">=</span> <span class="cn">FALSE</span>,</span>
<span>  report <span class="op">=</span> <span class="cn">TRUE</span>,</span>
<span>  report.dir <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  pathway_enrichment <span class="op">=</span> <span class="cn">FALSE</span>,</span>
<span>  pathway_kegg <span class="op">=</span> <span class="cn">FALSE</span>,</span>
<span>  kegg_organism <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  custom_pathways <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  out.dir <span class="op">=</span> <span class="cn">NULL</span></span>
<span><span class="op">)</span></span></code></pre></div>
    </div>

    <div class="section level2">
    <h2 id="arguments">Arguments<a class="anchor" aria-label="anchor" href="#arguments"></a></h2>


<dl><dt id="arg-se">se<a class="anchor" aria-label="anchor" href="#arg-se"></a></dt>
<dd><p><code>SummarizedExperiment</code> object, proteomics data parsed with <code><a href="prot.read_data.html">prot.read_data</a></code>.</p></dd>


<dt id="arg-imp-fun">imp_fun<a class="anchor" aria-label="anchor" href="#arg-imp-fun"></a></dt>
<dd><p>(Character string)  Function used for data imputation. "SampMin", "man", "bpca", "knn", "QRILC", "MLE", "MinDet", "MinProb", "min", "zero", "mixed", or "nbavg". See (<code><a href="prot.impute.html">prot.impute</a></code>) for details.</p></dd>


<dt id="arg-q">q<a class="anchor" aria-label="anchor" href="#arg-q"></a></dt>
<dd><p>(Numeric) q value for imputing missing values with method <code>imp_fun = 'MinProb'</code>.</p></dd>


<dt id="arg-knn-rowmax">knn.rowmax<a class="anchor" aria-label="anchor" href="#arg-knn-rowmax"></a></dt>
<dd><p>(Numeric) The maximum percent missing data allowed in any row for <code>imp_fun = 'knn'</code>. Default: 0.5.</p></dd>


<dt id="arg-type">type<a class="anchor" aria-label="anchor" href="#arg-type"></a></dt>
<dd><p>(Character string) Type of differential analysis to perform. "all" (contrast each condition with every other condition), "control" (contrast each condition to a defined control condition), "manual" (manually define selected conditions).</p></dd>


<dt id="arg-control">control<a class="anchor" aria-label="anchor" href="#arg-control"></a></dt>
<dd><p>(Character string) The name of the control condition if <code>type = control</code>.</p></dd>


<dt id="arg-contrast">contrast<a class="anchor" aria-label="anchor" href="#arg-contrast"></a></dt>
<dd><p>(Character string or vector of strings) Define the contrasts to be tested if <code>type = manual</code> in the form: <code>"ConditionA_vs_ConditionB"</code>, or <code>c("ConditionA_vs_ConditionC", "ConditionB_vs_ConditionC")</code>.</p></dd>


<dt id="arg-alpha">alpha<a class="anchor" aria-label="anchor" href="#arg-alpha"></a></dt>
<dd><p>(Numeric) Significance threshold for adjusted p values.</p></dd>


<dt id="arg-alpha-pathways">alpha_pathways<a class="anchor" aria-label="anchor" href="#arg-alpha-pathways"></a></dt>
<dd><p>(Numeric) Significance threshold for adjusted p values in pathway enrichment analysis.</p></dd>


<dt id="arg-lfc">lfc<a class="anchor" aria-label="anchor" href="#arg-lfc"></a></dt>
<dd><p>(Numeric) Relevance threshold for log2(fold change) values. Only proteins with a |log2(fold change)| value above <code>lfc</code> for a given contrast are considered "significant" (if they additionally fullfil the <code>alpha</code> criterion).</p></dd>


<dt id="arg-heatmap-show-all">heatmap.show_all<a class="anchor" aria-label="anchor" href="#arg-heatmap-show-all"></a></dt>
<dd><p>(Logical) Shall all samples be displayed in the heat map (<code>TRUE</code>) or only the samples contained in the defined <code>contrast</code> (<code>FALSE</code>)?</p></dd>


<dt id="arg-heatmap-kmeans">heatmap.kmeans<a class="anchor" aria-label="anchor" href="#arg-heatmap-kmeans"></a></dt>
<dd><p>(Logical) Shall the proteins be clustered in the heat map with the k-nearest neighbour method (<code>TRUE</code>) or not <code>FALSE</code>)?</p></dd>


<dt id="arg-k">k<a class="anchor" aria-label="anchor" href="#arg-k"></a></dt>
<dd><p>(Integer) Number of protein clusters in heat map if <code>heatmap.kmeans = TRUE</code>.</p></dd>


<dt id="arg-heatmap-col-limit">heatmap.col_limit<a class="anchor" aria-label="anchor" href="#arg-heatmap-col-limit"></a></dt>
<dd><p>(Integer) Define the outer breaks in the heat map legend. Example: if <code>heatmap.col_limit = 3</code>, the color scale will span from -3 to 3. Alls values below -3 will have the same color as -3, and all values above 3 will have the same color as 3.</p></dd>


<dt id="arg-heatmap-show-row-names">heatmap.show_row_names<a class="anchor" aria-label="anchor" href="#arg-heatmap-show-row-names"></a></dt>
<dd><p>(Logical) Show protein names in heat map (<code>TRUE</code>) or not <code>FALSE</code>).</p></dd>


<dt id="arg-heatmap-row-font-size">heatmap.row_font_size<a class="anchor" aria-label="anchor" href="#arg-heatmap-row-font-size"></a></dt>
<dd><p>(Numeric) Font size of protein names in heat maps if <code>heatmap.show_row_names = TRUE</code>.</p></dd>


<dt id="arg-volcano-add-names">volcano.add_names<a class="anchor" aria-label="anchor" href="#arg-volcano-add-names"></a></dt>
<dd><p>(Logical) Show protein names in volcano plots (<code>TRUE</code>) or not <code>FALSE</code>).</p></dd>


<dt id="arg-volcano-label-size">volcano.label_size<a class="anchor" aria-label="anchor" href="#arg-volcano-label-size"></a></dt>
<dd><p>(Numeric) Font size of protein names in volcano plots if <code>volcano.add_names = TRUE</code>.</p></dd>


<dt id="arg-volcano-adjusted">volcano.adjusted<a class="anchor" aria-label="anchor" href="#arg-volcano-adjusted"></a></dt>
<dd><p>(Logical) Shall adjusted p values be shown on the y axis of volcano plots (<code>TRUE</code>) or raw p values (<code>FALSE</code>)?.</p></dd>


<dt id="arg-plot">plot<a class="anchor" aria-label="anchor" href="#arg-plot"></a></dt>
<dd><p>(Logical) Show the generated plots in the <code>Plots</code> pane of RStudio (<code>TRUE</code>) or not <code>FALSE</code>).</p></dd>


<dt id="arg-export">export<a class="anchor" aria-label="anchor" href="#arg-export"></a></dt>
<dd><p>(Logical) Exported the generated plots as PNG and PDF files (<code>TRUE</code>) or not <code>FALSE</code>).</p></dd>


<dt id="arg-report">report<a class="anchor" aria-label="anchor" href="#arg-report"></a></dt>
<dd><p>(Logical) Render and export a report in PDF and HTML format that summarizes the results (<code>TRUE</code>) or not <code>FALSE</code>).</p></dd>


<dt id="arg-report-dir">report.dir<a class="anchor" aria-label="anchor" href="#arg-report-dir"></a></dt>
<dd><p>(Character string) Provide the name of or path to a folder into which the report will be saved.</p></dd>


<dt id="arg-pathway-enrichment">pathway_enrichment<a class="anchor" aria-label="anchor" href="#arg-pathway-enrichment"></a></dt>
<dd><p>(Logical) Perform pathway over-representation analysis for each tested contrast (<code>TRUE</code>) or not <code>FALSE</code>).</p></dd>


<dt id="arg-pathway-kegg">pathway_kegg<a class="anchor" aria-label="anchor" href="#arg-pathway-kegg"></a></dt>
<dd><p>(Logical) Perform pathway over-representation analysis with gene sets in the KEGG database (<code>TRUE</code>) or not <code>FALSE</code>).</p></dd>


<dt id="arg-kegg-organism">kegg_organism<a class="anchor" aria-label="anchor" href="#arg-kegg-organism"></a></dt>
<dd><p>(Character string) Identifier of the organism in the KEGG database (if <code>pathway_kegg = TRUE</code>)</p></dd>


<dt id="arg-custom-pathways">custom_pathways<a class="anchor" aria-label="anchor" href="#arg-custom-pathways"></a></dt>
<dd><p>(a R dataframe object) Data frame providing custom pathway annotations. The table must contain a "Pathway" column listing identified pathway in the studies organism, and an "Accession" column listing the proteins (or genes) each pathway is composed of. The <strong>Accession</strong> entries must match with protein <strong>IDs</strong>.</p></dd>


<dt id="arg-out-dir">out.dir<a class="anchor" aria-label="anchor" href="#arg-out-dir"></a></dt>
<dd><p>(Character string) absolute path to the location where result TXT files should be exported to.</p></dd>

</dl></div>
    <div class="section level2">
    <h2 id="value">Value<a class="anchor" aria-label="anchor" href="#value"></a></h2>
    <p>A list containing <code>SummarizedExperiment</code> object for every computation step of the workflow, a <code>pca</code> object, and lists of up- or down regulated pathways for each tested contrast and method (KEGG and/or custom).</p>
    </div>

  </main><aside class="col-md-3"><nav id="toc" aria-label="Table of contents"><h2>On this page</h2>
    </nav></aside></div>


    <footer><div class="pkgdown-footer-left">
  <p>Developed by Nicolas T. Wirth.</p>
</div>

<div class="pkgdown-footer-right">
  <p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.1.1.</p>
</div>

    </footer></div>





  </body></html>

