<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en"><head><meta http-equiv="Content-Type" content="text/html; charset=UTF-8"><meta charset="utf-8"><meta http-equiv="X-UA-Compatible" content="IE=edge"><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no"><title>RNA sequencing workflow — rna.workflow • VisomX</title><script src="../deps/jquery-3.6.0/jquery-3.6.0.min.js"></script><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no"><link href="../deps/bootstrap-5.3.1/bootstrap.min.css" rel="stylesheet"><script src="../deps/bootstrap-5.3.1/bootstrap.bundle.min.js"></script><link href="../deps/font-awesome-6.5.2/css/all.min.css" rel="stylesheet"><link href="../deps/font-awesome-6.5.2/css/v4-shims.min.css" rel="stylesheet"><script src="../deps/headroom-0.11.0/headroom.min.js"></script><script src="../deps/headroom-0.11.0/jQuery.headroom.min.js"></script><script src="../deps/bootstrap-toc-1.0.1/bootstrap-toc.min.js"></script><script src="../deps/clipboard.js-2.0.11/clipboard.min.js"></script><script src="../deps/search-1.0.0/autocomplete.jquery.min.js"></script><script src="../deps/search-1.0.0/fuse.min.js"></script><script src="../deps/search-1.0.0/mark.min.js"></script><!-- pkgdown --><script src="../pkgdown.js"></script><link href="../extra.css" rel="stylesheet"><meta property="og:title" content="RNA sequencing workflow — rna.workflow"><meta name="description" content="This function performs a complete RNA sequencing workflow, including imputation of missing values, normalization,
principal component analysis, differential expression analysis, and pathway analysis. The function also provides
several options for plotting, exporting plots, and creating a report."><meta property="og:description" content="This function performs a complete RNA sequencing workflow, including imputation of missing values, normalization,
principal component analysis, differential expression analysis, and pathway analysis. The function also provides
several options for plotting, exporting plots, and creating a report."></head><body>
    <a href="#main" class="visually-hidden-focusable">Skip to contents</a>


    <nav class="navbar navbar-expand-lg fixed-top bg-light" data-bs-theme="light" aria-label="Site navigation"><div class="container">

    <a class="navbar-brand me-2" href="../index.html">VisomX</a>

    <small class="nav-text text-muted me-auto" data-bs-toggle="tooltip" data-bs-placement="bottom" title="">0.0.94</small>


    <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbar" aria-controls="navbar" aria-expanded="false" aria-label="Toggle navigation">
      <span class="navbar-toggler-icon"></span>
    </button>

    <div id="navbar" class="collapse navbar-collapse ms-3">
      <ul class="navbar-nav me-auto"><li class="active nav-item"><a class="nav-link" href="../reference/index.html">Reference</a></li>
<li class="nav-item dropdown">
  <button class="nav-link dropdown-toggle" type="button" id="dropdown-articles" data-bs-toggle="dropdown" aria-expanded="false" aria-haspopup="true">Articles</button>
  <ul class="dropdown-menu" aria-labelledby="dropdown-articles"><li><a class="dropdown-item" href="../articles/vignette_proteomics.html">Reliable and extensive analysis of Proteomics data with package `VisomX`</a></li>
    <li><a class="dropdown-item" href="../articles/vignette_transcriptomics.html">Reliable and extensive analysis of transcriptomics data with package `VisomX`</a></li>
  </ul></li>
      </ul><ul class="navbar-nav"><li class="nav-item"><form class="form-inline" role="search">
 <input class="form-control" type="search" name="search-input" id="search-input" autocomplete="off" aria-label="Search site" placeholder="Search for" data-search-index="../search.json"></form></li>
<li class="nav-item"><a class="external-link nav-link" href="https://github.com/NicWir/VisomX/" aria-label="GitHub"><span class="fa fab fa-github fa-lg"></span></a></li>
      </ul></div>


  </div>
</nav><div class="container template-reference-topic">
<div class="row">
  <main id="main" class="col-md-9"><div class="page-header">

      <h1>RNA sequencing workflow</h1>
      <small class="dont-index">Source: <a href="https://github.com/NicWir/VisomX/blob/master/R/rna_workflow.R" class="external-link"><code>R/rna_workflow.R</code></a></small>
      <div class="d-none name"><code>rna.workflow.Rd</code></div>
    </div>

    <div class="ref-description section level2">
    <p>This function performs a complete RNA sequencing workflow, including imputation of missing values, normalization,
principal component analysis, differential expression analysis, and pathway analysis. The function also provides
several options for plotting, exporting plots, and creating a report.</p>
    </div>

    <div class="section level2">
    <h2 id="ref-usage">Usage<a class="anchor" aria-label="anchor" href="#ref-usage"></a></h2>
    <div class="sourceCode"><pre class="sourceCode r"><code><span><span class="fu">rna.workflow</span><span class="op">(</span></span>
<span>  <span class="va">se</span>,</span>
<span>  imp_fun <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"zero"</span>, <span class="st">"man"</span>, <span class="st">"bpca"</span>, <span class="st">"knn"</span>, <span class="st">"QRILC"</span>, <span class="st">"MLE"</span>, <span class="st">"MinDet"</span>, <span class="st">"MinProb"</span>, <span class="st">"min"</span>,</span>
<span>    <span class="st">"zero"</span>, <span class="st">"mixed"</span>, <span class="st">"nbavg"</span>, <span class="st">"SampMin"</span><span class="op">)</span>,</span>
<span>  q <span class="op">=</span> <span class="fl">0.01</span>,</span>
<span>  knn.rowmax <span class="op">=</span> <span class="fl">0.5</span>,</span>
<span>  type <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"all"</span>, <span class="st">"control"</span>, <span class="st">"manual"</span><span class="op">)</span>,</span>
<span>  design <span class="op">=</span> <span class="st">"~ 0 + condition"</span>,</span>
<span>  size.factors <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  altHypothesis <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"greaterAbs"</span>, <span class="st">"lessAbs"</span>, <span class="st">"greater"</span>, <span class="st">"less"</span><span class="op">)</span>,</span>
<span>  control <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  contrast <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  controlGenes <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  pAdjustMethod <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"IHW"</span>, <span class="st">"BH"</span><span class="op">)</span>,</span>
<span>  alpha <span class="op">=</span> <span class="fl">0.05</span>,</span>
<span>  alpha.independent <span class="op">=</span> <span class="fl">0.1</span>,</span>
<span>  alpha_pathways <span class="op">=</span> <span class="fl">0.1</span>,</span>
<span>  lfcShrink <span class="op">=</span> <span class="cn">TRUE</span>,</span>
<span>  shrink.method <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"apeglm"</span>, <span class="st">"ashr"</span>, <span class="st">"normal"</span><span class="op">)</span>,</span>
<span>  lfc <span class="op">=</span> <span class="fl">2</span>,</span>
<span>  heatmap.show_all <span class="op">=</span> <span class="cn">TRUE</span>,</span>
<span>  heatmap.kmeans <span class="op">=</span> <span class="cn">F</span>,</span>
<span>  k <span class="op">=</span> <span class="fl">6</span>,</span>
<span>  heatmap.col_limit <span class="op">=</span> <span class="cn">NA</span>,</span>
<span>  heatmap.show_row_names <span class="op">=</span> <span class="cn">TRUE</span>,</span>
<span>  heatmap.row_font_size <span class="op">=</span> <span class="fl">6</span>,</span>
<span>  volcano.add_names <span class="op">=</span> <span class="cn">FALSE</span>,</span>
<span>  volcano.label_size <span class="op">=</span> <span class="fl">2.5</span>,</span>
<span>  volcano.adjusted <span class="op">=</span> <span class="cn">TRUE</span>,</span>
<span>  plot <span class="op">=</span> <span class="cn">FALSE</span>,</span>
<span>  export <span class="op">=</span> <span class="cn">FALSE</span>,</span>
<span>  report <span class="op">=</span> <span class="cn">TRUE</span>,</span>
<span>  report.dir <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  pathway_enrichment <span class="op">=</span> <span class="cn">FALSE</span>,</span>
<span>  pathway_kegg <span class="op">=</span> <span class="cn">FALSE</span>,</span>
<span>  kegg_organism <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  custom_pathways <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  quiet <span class="op">=</span> <span class="cn">FALSE</span></span>
<span><span class="op">)</span></span></code></pre></div>
    </div>

    <div class="section level2">
    <h2 id="arguments">Arguments<a class="anchor" aria-label="anchor" href="#arguments"></a></h2>


<dl><dt id="arg-se">se<a class="anchor" aria-label="anchor" href="#arg-se"></a></dt>
<dd><p>A SummarizedExperiment object, generated with read_prot().</p></dd>


<dt id="arg-imp-fun">imp_fun<a class="anchor" aria-label="anchor" href="#arg-imp-fun"></a></dt>
<dd><p>(Character string)  Function used for data imputation. "SampMin", "man", "bpca", "knn", "QRILC", "MLE", "MinDet", "MinProb", "min", "zero", "mixed", or "nbavg". See (<code><a href="rna.impute.html">rna.impute</a></code>) for details.</p></dd>


<dt id="arg-q">q<a class="anchor" aria-label="anchor" href="#arg-q"></a></dt>
<dd><p>(Numeric) q value for imputing missing values with method <code>imp_fun = 'MinProb'</code>.</p></dd>


<dt id="arg-knn-rowmax">knn.rowmax<a class="anchor" aria-label="anchor" href="#arg-knn-rowmax"></a></dt>
<dd><p>(Numeric) The maximum percent missing data allowed in any row for <code>imp_fun = 'knn'</code>. Default: 0.5.</p></dd>


<dt id="arg-type">type<a class="anchor" aria-label="anchor" href="#arg-type"></a></dt>
<dd><p>(Character string) Type of differential analysis to perform. "all" (contrast each condition with every other condition), "control" (contrast each condition to a defined control condition), "manual" (manually define selected conditions).</p></dd>


<dt id="arg-design">design<a class="anchor" aria-label="anchor" href="#arg-design"></a></dt>
<dd><p>Formula for the design matrix.</p></dd>


<dt id="arg-size-factors">size.factors<a class="anchor" aria-label="anchor" href="#arg-size-factors"></a></dt>
<dd><p>Optional: Manually define size factors for normalization.</p></dd>


<dt id="arg-althypothesis">altHypothesis<a class="anchor" aria-label="anchor" href="#arg-althypothesis"></a></dt>
<dd><p>Specify those genes you are interested in finding. The test provides p values for the null hypothesis, the complement of the set defined by altHypothesis. For further details, see <code><a href="https://rdrr.io/pkg/DESeq2/man/results.html" class="external-link">results</a></code>.</p></dd>


<dt id="arg-control">control<a class="anchor" aria-label="anchor" href="#arg-control"></a></dt>
<dd><p>Control condition; required if type = "control".</p></dd>


<dt id="arg-contrast">contrast<a class="anchor" aria-label="anchor" href="#arg-contrast"></a></dt>
<dd><p>(String or vector of strings) Defined test(s) for differential analysis in the form "A_vs_B"; required if type = "manual".</p></dd>


<dt id="arg-controlgenes">controlGenes<a class="anchor" aria-label="anchor" href="#arg-controlgenes"></a></dt>
<dd><p>Specifying those genes to use for size factor estimation (e.g. housekeeping or spike-in genes).</p></dd>


<dt id="arg-padjustmethod">pAdjustMethod<a class="anchor" aria-label="anchor" href="#arg-padjustmethod"></a></dt>
<dd><p>Method for adjusting p values. Available options are "IHW" (Independent Hypothesis Weighting),"BH" (Benjamini-Hochberg).</p></dd>


<dt id="arg-alpha">alpha<a class="anchor" aria-label="anchor" href="#arg-alpha"></a></dt>
<dd><p>Significance threshold for adjusted p values.</p></dd>


<dt id="arg-alpha-independent">alpha.independent<a class="anchor" aria-label="anchor" href="#arg-alpha-independent"></a></dt>
<dd><p>Adjusted p value threshold for independent filtering or NULL. If the adjusted p-value cutoff (FDR) will be a value other than 0.1, alpha should be set to that value.</p></dd>


<dt id="arg-alpha-pathways">alpha_pathways<a class="anchor" aria-label="anchor" href="#arg-alpha-pathways"></a></dt>
<dd><p>Significance threshold for pathway analysis.</p></dd>


<dt id="arg-lfcshrink">lfcShrink<a class="anchor" aria-label="anchor" href="#arg-lfcshrink"></a></dt>
<dd><p>Use shrinkage to calculate log2 fold change values.</p></dd>


<dt id="arg-shrink-method">shrink.method<a class="anchor" aria-label="anchor" href="#arg-shrink-method"></a></dt>
<dd><p>Method for shrinkage. Available options are "apeglm", "ashr", "normal". See <code><a href="https://rdrr.io/pkg/DESeq2/man/lfcShrink.html" class="external-link">lfcShrink</a></code> for details.</p></dd>


<dt id="arg-lfc">lfc<a class="anchor" aria-label="anchor" href="#arg-lfc"></a></dt>
<dd><p>Relevance threshold for absolute log2(fold change) values. Used to filter unshrunken lfc values
or in shrinkage method "apeglm" or "normal".</p></dd>


<dt id="arg-heatmap-show-all">heatmap.show_all<a class="anchor" aria-label="anchor" href="#arg-heatmap-show-all"></a></dt>
<dd><p>Shall all samples be displayed in the heatmap or only the samples contained in the defined
"contrast"? (only applicable for type = "manual")</p></dd>


<dt id="arg-heatmap-kmeans">heatmap.kmeans<a class="anchor" aria-label="anchor" href="#arg-heatmap-kmeans"></a></dt>
<dd><p>Shall the proteins be clustered in the heat map?</p></dd>


<dt id="arg-k">k<a class="anchor" aria-label="anchor" href="#arg-k"></a></dt>
<dd><p>Number of protein clusters in heat map if kmeans = TRUE.</p></dd>


<dt id="arg-heatmap-col-limit">heatmap.col_limit<a class="anchor" aria-label="anchor" href="#arg-heatmap-col-limit"></a></dt>
<dd><p>Define the breaks in the heat map legends.</p></dd>


<dt id="arg-heatmap-show-row-names">heatmap.show_row_names<a class="anchor" aria-label="anchor" href="#arg-heatmap-show-row-names"></a></dt>
<dd><p>Show protein names in heat map?</p></dd>


<dt id="arg-heatmap-row-font-size">heatmap.row_font_size<a class="anchor" aria-label="anchor" href="#arg-heatmap-row-font-size"></a></dt>
<dd><p>Font size of protein names if show_row_names = TRUE.</p></dd>


<dt id="arg-volcano-add-names">volcano.add_names<a class="anchor" aria-label="anchor" href="#arg-volcano-add-names"></a></dt>
<dd><p>Display names next to symbols in volcano plot.</p></dd>


<dt id="arg-volcano-label-size">volcano.label_size<a class="anchor" aria-label="anchor" href="#arg-volcano-label-size"></a></dt>
<dd><p>Size of labels in volcano plot.</p></dd>


<dt id="arg-volcano-adjusted">volcano.adjusted<a class="anchor" aria-label="anchor" href="#arg-volcano-adjusted"></a></dt>
<dd><p>Display adjusted p-values on y axis of volcano plot?</p></dd>


<dt id="arg-plot">plot<a class="anchor" aria-label="anchor" href="#arg-plot"></a></dt>
<dd><p>Shall plots be returned in the Plots pane?</p></dd>


<dt id="arg-export">export<a class="anchor" aria-label="anchor" href="#arg-export"></a></dt>
<dd><p>Shall plots be exported as PDF and PNG files?</p></dd>


<dt id="arg-report">report<a class="anchor" aria-label="anchor" href="#arg-report"></a></dt>
<dd><p>Shall a report (HTML and PDF) be created?</p></dd>


<dt id="arg-report-dir">report.dir<a class="anchor" aria-label="anchor" href="#arg-report-dir"></a></dt>
<dd><p>Folder name for created report (if report = TRUE)</p></dd>


<dt id="arg-pathway-enrichment">pathway_enrichment<a class="anchor" aria-label="anchor" href="#arg-pathway-enrichment"></a></dt>
<dd><p>Perform pathway over-representation analysis for each tested contrast</p></dd>


<dt id="arg-pathway-kegg">pathway_kegg<a class="anchor" aria-label="anchor" href="#arg-pathway-kegg"></a></dt>
<dd><p>Perform pathway over-representation analysis with gene sets in the KEGG database</p></dd>


<dt id="arg-kegg-organism">kegg_organism<a class="anchor" aria-label="anchor" href="#arg-kegg-organism"></a></dt>
<dd><p>Name of the organism in the KEGG database (if 'pathway_kegg = TRUE')</p></dd>


<dt id="arg-custom-pathways">custom_pathways<a class="anchor" aria-label="anchor" href="#arg-custom-pathways"></a></dt>
<dd><p>Dataframe providing custom pathway annotations.</p></dd>


<dt id="arg-quiet">quiet<a class="anchor" aria-label="anchor" href="#arg-quiet"></a></dt>
<dd><p>Suppress messages and warnings.</p></dd>

</dl></div>
    <div class="section level2">
    <h2 id="value">Value<a class="anchor" aria-label="anchor" href="#value"></a></h2>
    <p>The function returns a SummarizedExperiment object with added columns for log2 fold change, p-values and adjusted p-values for each comparison.
It also includes a column for significant genes for each comparison and a column for significant genes overall.
Additionally, the function generates various plots and a report (if specified).</p>
    </div>

  </main><aside class="col-md-3"><nav id="toc" aria-label="Table of contents"><h2>On this page</h2>
    </nav></aside></div>


    <footer><div class="pkgdown-footer-left">
  <p>Developed by Nicolas T. Wirth.</p>
</div>

<div class="pkgdown-footer-right">
  <p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.1.2.</p>
</div>

    </footer></div>





  </body></html>

