<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en"><head><meta http-equiv="Content-Type" content="text/html; charset=UTF-8"><meta charset="utf-8"><meta http-equiv="X-UA-Compatible" content="IE=edge"><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no"><title>Plot bar plots for gene abundance or fold change — rna.plot_bar • VisomX</title><script src="../deps/jquery-3.6.0/jquery-3.6.0.min.js"></script><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no"><link href="../deps/bootstrap-5.3.1/bootstrap.min.css" rel="stylesheet"><script src="../deps/bootstrap-5.3.1/bootstrap.bundle.min.js"></script><link href="../deps/font-awesome-6.5.2/css/all.min.css" rel="stylesheet"><link href="../deps/font-awesome-6.5.2/css/v4-shims.min.css" rel="stylesheet"><script src="../deps/headroom-0.11.0/headroom.min.js"></script><script src="../deps/headroom-0.11.0/jQuery.headroom.min.js"></script><script src="../deps/bootstrap-toc-1.0.1/bootstrap-toc.min.js"></script><script src="../deps/clipboard.js-2.0.11/clipboard.min.js"></script><script src="../deps/search-1.0.0/autocomplete.jquery.min.js"></script><script src="../deps/search-1.0.0/fuse.min.js"></script><script src="../deps/search-1.0.0/mark.min.js"></script><!-- pkgdown --><script src="../pkgdown.js"></script><link href="../extra.css" rel="stylesheet"><meta property="og:title" content="Plot bar plots for gene abundance or fold change — rna.plot_bar"><meta name="description" content="This function generates bar plots for gene abundance or fold change, normalized by either a reference gene or centered to the mean. The user can also plot log2 intensities, contrasts, or raw counts."><meta property="og:description" content="This function generates bar plots for gene abundance or fold change, normalized by either a reference gene or centered to the mean. The user can also plot log2 intensities, contrasts, or raw counts."></head><body>
    <a href="#main" class="visually-hidden-focusable">Skip to contents</a>


    <nav class="navbar navbar-expand-lg fixed-top bg-light" data-bs-theme="light" aria-label="Site navigation"><div class="container">

    <a class="navbar-brand me-2" href="../index.html">VisomX</a>

    <small class="nav-text text-muted me-auto" data-bs-toggle="tooltip" data-bs-placement="bottom" title="">0.0.94</small>


    <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbar" aria-controls="navbar" aria-expanded="false" aria-label="Toggle navigation">
      <span class="navbar-toggler-icon"></span>
    </button>

    <div id="navbar" class="collapse navbar-collapse ms-3">
      <ul class="navbar-nav me-auto"><li class="active nav-item"><a class="nav-link" href="../reference/index.html">Reference</a></li>
<li class="nav-item dropdown">
  <button class="nav-link dropdown-toggle" type="button" id="dropdown-articles" data-bs-toggle="dropdown" aria-expanded="false" aria-haspopup="true">Articles</button>
  <ul class="dropdown-menu" aria-labelledby="dropdown-articles"><li><a class="dropdown-item" href="../articles/vignette_proteomics.html">Reliable and extensive analysis of Proteomics data with package `VisomX`</a></li>
    <li><a class="dropdown-item" href="../articles/vignette_transcriptomics.html">Reliable and extensive analysis of transcriptomics data with package `VisomX`</a></li>
  </ul></li>
      </ul><ul class="navbar-nav"><li class="nav-item"><form class="form-inline" role="search">
 <input class="form-control" type="search" name="search-input" id="search-input" autocomplete="off" aria-label="Search site" placeholder="Search for" data-search-index="../search.json"></form></li>
<li class="nav-item"><a class="external-link nav-link" href="https://github.com/NicWir/VisomX/" aria-label="GitHub"><span class="fa fab fa-github fa-lg"></span></a></li>
      </ul></div>


  </div>
</nav><div class="container template-reference-topic">
<div class="row">
  <main id="main" class="col-md-9"><div class="page-header">

      <h1>Plot bar plots for gene abundance or fold change</h1>
      <small class="dont-index">Source: <a href="https://github.com/NicWir/VisomX/blob/master/R/transcriptomics_plots.R" class="external-link"><code>R/transcriptomics_plots.R</code></a></small>
      <div class="d-none name"><code>rna.plot_bar.Rd</code></div>
    </div>

    <div class="ref-description section level2">
    <p>This function generates bar plots for gene abundance or fold change, normalized by either a reference gene or centered to the mean. The user can also plot log2 intensities, contrasts, or raw counts.</p>
    </div>

    <div class="section level2">
    <h2 id="ref-usage">Usage<a class="anchor" aria-label="anchor" href="#ref-usage"></a></h2>
    <div class="sourceCode"><pre class="sourceCode r"><code><span><span class="fu">rna.plot_bar</span><span class="op">(</span></span>
<span>  <span class="va">dds</span>,</span>
<span>  <span class="va">genes</span>,</span>
<span>  type <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"contrast"</span>, <span class="st">"log_centered"</span>, <span class="st">"centered"</span>, <span class="st">"raw"</span>, <span class="st">"log2"</span><span class="op">)</span>,</span>
<span>  shrunken.lfc <span class="op">=</span> <span class="cn">TRUE</span>,</span>
<span>  contrast <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  conditions <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  col.id <span class="op">=</span> <span class="st">"ID"</span>,</span>
<span>  name_table <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  match.id <span class="op">=</span> <span class="st">"Accession"</span>,</span>
<span>  match.name <span class="op">=</span> <span class="st">"Name"</span>,</span>
<span>  convert_name <span class="op">=</span> <span class="cn">FALSE</span>,</span>
<span>  shape.size <span class="op">=</span> <span class="fl">2.5</span>,</span>
<span>  collage <span class="op">=</span> <span class="cn">TRUE</span>,</span>
<span>  plot <span class="op">=</span> <span class="cn">TRUE</span>,</span>
<span>  export <span class="op">=</span> <span class="cn">TRUE</span>,</span>
<span>  export.nm <span class="op">=</span> <span class="cn">NULL</span></span>
<span><span class="op">)</span></span></code></pre></div>
    </div>

    <div class="section level2">
    <h2 id="arguments">Arguments<a class="anchor" aria-label="anchor" href="#arguments"></a></h2>


<dl><dt id="arg-dds">dds<a class="anchor" aria-label="anchor" href="#arg-dds"></a></dt>
<dd><p>A DESeqDataSet object containing the data.</p></dd>


<dt id="arg-genes">genes<a class="anchor" aria-label="anchor" href="#arg-genes"></a></dt>
<dd><p>A vector of gene names or IDs.</p></dd>


<dt id="arg-type">type<a class="anchor" aria-label="anchor" href="#arg-type"></a></dt>
<dd><p>Type of plot: "contrast", "log_centered", "centered", "raw", or "log2".</p><ul><li><p>"log_centered": log2-transformed intensities centered to their mean.</p></li>
<li><p>"centered": raw intensities centered to their mean.</p></li>
<li><p>"contrast": log2 fold changes for specified contrasts.</p></li>
<li><p>"raw": raw (or normalized) read counts with mean and 95% CI.</p></li>
<li><p>"log2": log2-transformed intensities with mean and 95% CI.</p></li>
</ul></dd>


<dt id="arg-shrunken-lfc">shrunken.lfc<a class="anchor" aria-label="anchor" href="#arg-shrunken-lfc"></a></dt>
<dd><p>A logical indicating if shrunken log2 fold changes should be used if <code>type = "contrast"</code>. Default is TRUE.</p></dd>


<dt id="arg-contrast">contrast<a class="anchor" aria-label="anchor" href="#arg-contrast"></a></dt>
<dd><p>Name of the contrast for plotting when type is "contrast".</p></dd>


<dt id="arg-conditions">conditions<a class="anchor" aria-label="anchor" href="#arg-conditions"></a></dt>
<dd><p>Character vector of condition names to be plotted.</p></dd>


<dt id="arg-col-id">col.id<a class="anchor" aria-label="anchor" href="#arg-col-id"></a></dt>
<dd><p>Name of the column with gene IDs in <code>dds</code>.</p></dd>


<dt id="arg-name-table">name_table<a class="anchor" aria-label="anchor" href="#arg-name-table"></a></dt>
<dd><p>A data frame containing the mapping of gene names to reference names.</p></dd>


<dt id="arg-match-id">match.id<a class="anchor" aria-label="anchor" href="#arg-match-id"></a></dt>
<dd><p>The name of a column in <code>name_table</code> to find matches with <code>col.id</code>.</p></dd>


<dt id="arg-match-name">match.name<a class="anchor" aria-label="anchor" href="#arg-match-name"></a></dt>
<dd><p>The name of a column in <code>name_table</code> to assign new names based on matches of <code>col.id</code> and <code>match.id</code>.</p></dd>


<dt id="arg-convert-name">convert_name<a class="anchor" aria-label="anchor" href="#arg-convert-name"></a></dt>
<dd><p>Logical indicating whether to convert gene names based on <code>name_table</code>.</p></dd>


<dt id="arg-shape-size">shape.size<a class="anchor" aria-label="anchor" href="#arg-shape-size"></a></dt>
<dd><p>Numeric; replicate symbol size in the plot.</p></dd>


<dt id="arg-collage">collage<a class="anchor" aria-label="anchor" href="#arg-collage"></a></dt>
<dd><p>Logical; combine multiple gene plots into collages of up to 8. Default TRUE.</p></dd>


<dt id="arg-plot">plot<a class="anchor" aria-label="anchor" href="#arg-plot"></a></dt>
<dd><p>Logical; print the plot if TRUE or return data frame if FALSE. Default TRUE.</p></dd>


<dt id="arg-export">export<a class="anchor" aria-label="anchor" href="#arg-export"></a></dt>
<dd><p>Logical; export plots as PDF/PNG if TRUE. Default TRUE.</p></dd>


<dt id="arg-export-nm">export.nm<a class="anchor" aria-label="anchor" href="#arg-export-nm"></a></dt>
<dd><p>Base name for exported files if <code>export = TRUE</code>.</p></dd>

</dl></div>
    <div class="section level2">
    <h2 id="value">Value<a class="anchor" aria-label="anchor" href="#value"></a></h2>
    <p>If <code>plot = FALSE</code> and length(genes) &lt;= 8, returns a data frame of values and 95% CIs.</p>
    </div>

  </main><aside class="col-md-3"><nav id="toc" aria-label="Table of contents"><h2>On this page</h2>
    </nav></aside></div>


    <footer><div class="pkgdown-footer-left">
  <p>Developed by Nicolas T. Wirth.</p>
</div>

<div class="pkgdown-footer-right">
  <p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.1.3.</p>
</div>

    </footer></div>





  </body></html>

