<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en"><head><meta http-equiv="Content-Type" content="text/html; charset=UTF-8"><meta charset="utf-8"><meta http-equiv="X-UA-Compatible" content="IE=edge"><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no"><meta name="description" content="prot.read_data takes a table file containing proteomics data and filters proteins based on missing values (see prot.filter_missing) or a given relative standard deviation threshold, and creates a SummarizedExperiment object."><title>Read proteomics data in table format and create SummarizedExperiment — prot.read_data • VisomX</title><script src="../deps/jquery-3.6.0/jquery-3.6.0.min.js"></script><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no"><link href="../deps/bootstrap-5.3.1/bootstrap.min.css" rel="stylesheet"><script src="../deps/bootstrap-5.3.1/bootstrap.bundle.min.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous"><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous"><!-- bootstrap-toc --><script src="https://cdn.jsdelivr.net/gh/afeld/bootstrap-toc@v1.0.1/dist/bootstrap-toc.min.js" integrity="sha256-4veVQbu7//Lk5TSmc7YV48MxtMy98e26cf5MrgZYnwo=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- search --><script src="https://cdnjs.cloudflare.com/ajax/libs/fuse.js/6.4.6/fuse.js" integrity="sha512-zv6Ywkjyktsohkbp9bb45V6tEMoWhzFzXis+LrMehmJZZSys19Yxf1dopHx7WzIKxr5tK2dVcYmaCk2uqdjF4A==" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/autocomplete.js/0.38.0/autocomplete.jquery.min.js" integrity="sha512-GU9ayf+66Xx2TmpxqJpliWbT5PiGYxpaG8rfnBEk1LL8l1KGkRShhngwdXK1UgqhAzWpZHSiYPc09/NwDQIGyg==" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mark.js/8.11.1/mark.min.js" integrity="sha512-5CYOlHXGh6QpOFA/TeTylKLWfB3ftPsde7AnmhuitiTX4K5SqCLBeKro6sPS8ilsz1Q4NRx3v8Ko2IBiszzdww==" crossorigin="anonymous"></script><!-- pkgdown --><script src="../pkgdown.js"></script><link href="../extra.css" rel="stylesheet"><meta property="og:title" content="Read proteomics data in table format and create SummarizedExperiment — prot.read_data"><meta property="og:description" content="prot.read_data takes a table file containing proteomics data and filters proteins based on missing values (see prot.filter_missing) or a given relative standard deviation threshold, and creates a SummarizedExperiment object."><!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]--></head><body>
    <a href="#main" class="visually-hidden-focusable">Skip to contents</a>
    

    <nav class="navbar fixed-top navbar-light navbar-expand-lg bg-light"><div class="container">
    
    <a class="navbar-brand me-2" href="../index.html">VisomX</a>

    <small class="nav-text text-muted me-auto" data-bs-toggle="tooltip" data-bs-placement="bottom" title="">0.0.93</small>

    
    <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbar" aria-controls="navbar" aria-expanded="false" aria-label="Toggle navigation">
      <span class="navbar-toggler-icon"></span>
    </button>

    <div id="navbar" class="collapse navbar-collapse ms-3">
      <ul class="navbar-nav me-auto"><li class="active nav-item">
  <a class="nav-link" href="../reference/index.html">Reference</a>
</li>
<li class="nav-item dropdown">
  <a href="#" class="nav-link dropdown-toggle" data-bs-toggle="dropdown" role="button" aria-expanded="false" aria-haspopup="true" id="dropdown-articles">Articles</a>
  <div class="dropdown-menu" aria-labelledby="dropdown-articles">
    <a class="dropdown-item" href="../articles/vignette_proteomics.html">Reliable and extensive analysis of Proteomics data with package `VisomX`</a>
    <a class="dropdown-item" href="../articles/vignette_transcriptomics.html">Reliable and extensive analysis of transcriptomics data with package `VisomX`</a>
  </div>
</li>
      </ul><form class="form-inline my-2 my-lg-0" role="search">
        <input type="search" class="form-control me-sm-2" aria-label="Toggle navigation" name="search-input" data-search-index="../search.json" id="search-input" placeholder="Search for" autocomplete="off"></form>

      <ul class="navbar-nav"><li class="nav-item">
  <a class="external-link nav-link" href="https://github.com/NicWir/VisomX/" aria-label="github">
    <span class="fab fa fab fa-github fa-lg"></span>
     
  </a>
</li>
      </ul></div>

    
  </div>
</nav><div class="container template-reference-topic">
<div class="row">
  <main id="main" class="col-md-9"><div class="page-header">
      <img src="" class="logo" alt=""><h1>Read proteomics data in table format and create SummarizedExperiment</h1>
      <small class="dont-index">Source: <a href="https://github.com/NicWir/VisomX/blob/HEAD/R/proteomics_computation.R" class="external-link"><code>R/proteomics_computation.R</code></a></small>
      <div class="d-none name"><code>prot.read_data.Rd</code></div>
    </div>

    <div class="ref-description section level2">
    <p><code>prot.read_data</code> takes a table file containing proteomics data and filters proteins based on missing values (see <code><a href="prot.filter_missing.html">prot.filter_missing</a></code>) or a given relative standard deviation threshold, and creates a <code>SummarizedExperiment</code> object.</p>
    </div>

    <div class="section level2">
    <h2 id="ref-usage">Usage<a class="anchor" aria-label="anchor" href="#ref-usage"></a></h2>
    <div class="sourceCode"><pre class="sourceCode r"><code><span><span class="fu">prot.read_data</span><span class="op">(</span></span>
<span>  data <span class="op">=</span> <span class="st">"dat_prot.csv"</span>,</span>
<span>  expdesign <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  csvsep <span class="op">=</span> <span class="st">";"</span>,</span>
<span>  dec <span class="op">=</span> <span class="st">"."</span>,</span>
<span>  na.strings <span class="op">=</span> <span class="st">""</span>,</span>
<span>  sheet <span class="op">=</span> <span class="fl">1</span>,</span>
<span>  filter <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"Reverse"</span>, <span class="st">"Potential contaminant"</span><span class="op">)</span>,</span>
<span>  rsd_thresh <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  name <span class="op">=</span> <span class="st">"Gene Symbol"</span>,</span>
<span>  id <span class="op">=</span> <span class="st">"Ensembl Gene ID"</span>,</span>
<span>  pfx <span class="op">=</span> <span class="st">"abundances."</span>,</span>
<span>  filt_type <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"condition"</span>, <span class="st">"complete"</span>, <span class="st">"fraction"</span>, <span class="cn">NULL</span><span class="op">)</span>,</span>
<span>  filt_thr <span class="op">=</span> <span class="fl">3</span>,</span>
<span>  filt_min <span class="op">=</span> <span class="cn">NULL</span></span>
<span><span class="op">)</span></span></code></pre></div>
    </div>

    <div class="section level2">
    <h2 id="arguments">Arguments<a class="anchor" aria-label="anchor" href="#arguments"></a></h2>
    <dl><dt>data</dt>
<dd><p>An R dataframe object or a table file with extension '.xlsx', '.xls', '.csv', '.tsv', or '.txt' containing proteomics data.
The table must contain:</p><ol><li><p>A column with protein IDs (e.g., accession numbers). The header of this column is provided as argument <code>id</code>.</p></li>
<li><p>A column with protein names. The header of this column is provided as argument <code>name</code>.</p></li>
<li><p>X columns containing abundance values for X samples. The column headers must have a prefix (e.g., "abundances.") that is provided as argument <code>pfx</code>. Replicates are identified by identical column headers followed by an underscore and the replicate number (e.g., "abundances.ConditionA_1", "abundances.ConditionA_2", "abundances.ConditionA_3", ...).</p></li>
</ol></dd>


<dt>expdesign</dt>
<dd><p>(<em>optional, if made previously</em>) An R dataframe object or a table file containing the columns 'label', 'condition', and 'replicate' with label = "condition_replicate". If <code>NULL</code>, an experimental design table will be created automatically.</p></dd>


<dt>csvsep</dt>
<dd><p>(Character string) separator used in CSV files (ignored for other file types). Default: <code>";"</code></p></dd>


<dt>dec</dt>
<dd><p>(Character string) decimal separator used in CSV, TSV or TXT files (ignored for other file types). Default: <code>"."</code></p></dd>


<dt>na.strings</dt>
<dd><p>A character vector of strings which are to be interpreted as NA values.</p></dd>


<dt>sheet</dt>
<dd><p>(Integer or Character string) Number or name of the sheet with proteomics data in XLS or XLSX files (<em>optional</em>).</p></dd>


<dt>filter</dt>
<dd><p>(Character string or vector of strings) Provide the header of a column containing "+" or "-" to indicate if proteins should be discarded or kept, respectively.</p></dd>


<dt>rsd_thresh</dt>
<dd><p>(Numeric, optional) Provide a relative standard deviation (RSD) threshold <strong>in %</strong> for proteins. The RSD is calculated for each condition and if the maximum RSD value determined for a given protein exceeds <code>rsd_thresh</code>, the protein is discarded. The RSD filter is applied <strong>before</strong> further missing value filters based on the three <code>filt_</code> arguments.</p></dd>


<dt>name</dt>
<dd><p>(Character string) Provide the header of the column containing protein names.</p></dd>


<dt>id</dt>
<dd><p>(Character string) Provide the header of the column containing protein IDs</p></dd>


<dt>pfx</dt>
<dd><p>(Character string) Provide the common prefix for headers containing abundance values (e.g., "abundances.").</p></dd>


<dt>filt_type</dt>
<dd><p>(Character string) "complete", "condition" or "fraction", Sets the type of filtering applied. "complete" will only keep proteins with valid values in all samples. "condition" will keep proteins that have a maximum of <code>filt_thr</code> missing values in at least one condition. "fraction" will keep proteins that have a <code>filt_min</code> fraction of valid values in all samples.</p></dd>


<dt>filt_thr</dt>
<dd><p>(Integer) Sets the threshold for the allowed number of missing values in at least one condition if <code>filt_type = "condition"</code>. In other words: "keep proteins that have a maximum of 'filt_thr' missing values in at least one condition."</p></dd>


<dt>filt_min</dt>
<dd><p>(Numeric) Sets the threshold for the minimum fraction of valid values allowed for any protein if <code>filt_type = "fraction"</code>.</p></dd>

</dl></div>
    <div class="section level2">
    <h2 id="value">Value<a class="anchor" aria-label="anchor" href="#value"></a></h2>
    

<p>A filtered SummarizedExperiment object.</p>
    </div>

  </main><aside class="col-md-3"><nav id="toc"><h2>On this page</h2>
    </nav></aside></div>


    <footer><div class="pkgdown-footer-left">
  <p></p><p>Developed by Nicolas T. Wirth.</p>
</div>

<div class="pkgdown-footer-right">
  <p></p><p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.0.7.</p>
</div>

    </footer></div>

  

  

  </body></html>

