<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<title>Reliable and extensive analysis of Proteomics data with package `VisomX` • VisomX</title>
<script src="../deps/jquery-3.6.0/jquery-3.6.0.min.js"></script><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<link href="../deps/bootstrap-5.3.1/bootstrap.min.css" rel="stylesheet">
<script src="../deps/bootstrap-5.3.1/bootstrap.bundle.min.js"></script><link href="../deps/font-awesome-6.5.2/css/all.min.css" rel="stylesheet">
<link href="../deps/font-awesome-6.5.2/css/v4-shims.min.css" rel="stylesheet">
<script src="../deps/headroom-0.11.0/headroom.min.js"></script><script src="../deps/headroom-0.11.0/jQuery.headroom.min.js"></script><script src="../deps/bootstrap-toc-1.0.1/bootstrap-toc.min.js"></script><script src="../deps/clipboard.js-2.0.11/clipboard.min.js"></script><script src="../deps/search-1.0.0/autocomplete.jquery.min.js"></script><script src="../deps/search-1.0.0/fuse.min.js"></script><script src="../deps/search-1.0.0/mark.min.js"></script><!-- pkgdown --><script src="../pkgdown.js"></script><link href="../extra.css" rel="stylesheet">
<meta property="og:title" content="Reliable and extensive analysis of Proteomics data with package `VisomX`">
</head>
<body>
    <a href="#main" class="visually-hidden-focusable">Skip to contents</a>


    <nav class="navbar navbar-expand-lg fixed-top bg-light" data-bs-theme="light" aria-label="Site navigation"><div class="container">

    <a class="navbar-brand me-2" href="../index.html">VisomX</a>

    <small class="nav-text text-muted me-auto" data-bs-toggle="tooltip" data-bs-placement="bottom" title="">0.0.94</small>


    <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbar" aria-controls="navbar" aria-expanded="false" aria-label="Toggle navigation">
      <span class="navbar-toggler-icon"></span>
    </button>

    <div id="navbar" class="collapse navbar-collapse ms-3">
      <ul class="navbar-nav me-auto">
<li class="nav-item"><a class="nav-link" href="../reference/index.html">Reference</a></li>
<li class="active nav-item dropdown">
  <button class="nav-link dropdown-toggle" type="button" id="dropdown-articles" data-bs-toggle="dropdown" aria-expanded="false" aria-haspopup="true">Articles</button>
  <ul class="dropdown-menu" aria-labelledby="dropdown-articles">
<li><a class="dropdown-item" href="../articles/vignette_proteomics.html">Reliable and extensive analysis of Proteomics data with package `VisomX`</a></li>
    <li><a class="dropdown-item" href="../articles/vignette_transcriptomics.html">Reliable and extensive analysis of transcriptomics data with package `VisomX`</a></li>
  </ul>
</li>
      </ul>
<ul class="navbar-nav">
<li class="nav-item"><form class="form-inline" role="search">
 <input class="form-control" type="search" name="search-input" id="search-input" autocomplete="off" aria-label="Search site" placeholder="Search for" data-search-index="../search.json">
</form></li>
<li class="nav-item"><a class="external-link nav-link" href="https://github.com/NicWir/VisomX/" aria-label="GitHub"><span class="fa fab fa-github fa-lg"></span></a></li>
      </ul>
</div>


  </div>
</nav><div class="container template-article">




<div class="row">
  <main id="main" class="col-md-9"><div class="page-header">

      <h1>Reliable and extensive analysis of Proteomics data with package `VisomX`</h1>
                        <h4 data-toc-skip class="author">Nicolas T.
Wirth</h4>
            
            <h4 data-toc-skip class="date">2022-01-31</h4>
      
      <small class="dont-index">Source: <a href="https://github.com/NicWir/VisomX/blob/master/vignettes/vignette_proteomics.Rmd" class="external-link"><code>vignettes/vignette_proteomics.Rmd</code></a></small>
      <div class="d-none name"><code>vignette_proteomics.Rmd</code></div>
    </div>

    
    
<p style="text-align:center;">
<img src="https://user-images.githubusercontent.com/43075653/166063689-45b4e34e-7129-42c3-b109-3a34afdc24e2.png" class="center" width="433" height="178"></p>
<div class="section level2">
<h2 id="introduction">Introduction<a class="anchor" aria-label="anchor" href="#introduction"></a>
</h2>
<p>The aim of this <code>VisomX</code>’s proteomics section is to
provide an integrated analysis workflow for robust and reproducible
analysis of mass spectrometry proteomics data for differential protein
expression or differential enrichment. This section is build on the
foundation of the R package <code>DEP</code> <span class="citation">(<a href="#ref-zhang2018">Zhang <em>et al.</em>, 2018</a>)</span>.
<code>VisomX</code> was designed to be usable with minimal prior
knowledge of the R programming language or programming in general. You
will need to be familiar with the idea of running commands from a
console or writing basic scripts. For R beginners, <a href="https://moderndive.netlify.app/1-getting-started.html" class="external-link">this</a> is
a great starting point, there are some good resources <a href="https://education.rstudio.com/learn/beginner/" class="external-link">here</a> and we
suggest using the <a href="https://posit.co/products/open-source/rstudio/" class="external-link">RStudio
application</a>. It provides an environment for writing and running R
code.</p>
<p>With consideration for <code>R</code> novices, <code>VisomX</code>
establishes a framework in which a complete, detailed proteomics data
analysis can be performed in two simple steps:</p>
<ol style="list-style-type: decimal">
<li><p><em>Read data</em> from a tabular input with protein abundances,
as generated by quantitative analysis software of raw mass spectrometry
data, and apply several filters.</p></li>
<li><p><em>Run workflow</em>, including missing value imputation,
normalization, log2 transformation, statistical analysis, differential
expression analysis, pathway enrichment analysis, and rendering of a
report that summarizes the results.</p></li>
</ol>
<p>All computational results of a workflow are stored in a data
container (list) and can be visualized by passing them to a set of
dedicated plotting functions.</p>
<div style="page-break-after: always;"></div>
</div>
<div class="section level2">
<h2 id="installation">Installation<a class="anchor" aria-label="anchor" href="#installation"></a>
</h2>
<p>Install the most current version with package
<code>devtools</code>:</p>
<div class="sourceCode" id="cb1"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/utils/install.packages.html" class="external-link">install.packages</a></span><span class="op">(</span><span class="st">"BiocManager"</span><span class="op">)</span></span>
<span></span>
<span><span class="fu">BiocManager</span><span class="fu">::</span><span class="fu"><a href="https://bioconductor.github.io/BiocManager/reference/install.html" class="external-link">install</a></span><span class="op">(</span><span class="st">"NicWir/VisomX"</span><span class="op">)</span></span></code></pre></div>
<p>Load the package:</p>
<div class="sourceCode" id="cb2"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Load the VisomX package</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://github.com/NicWir/VisomX" class="external-link">VisomX</a></span><span class="op">)</span></span></code></pre></div>
</div>
<div class="section level2">
<h2 id="the-test-dataset">The Test Dataset<a class="anchor" aria-label="anchor" href="#the-test-dataset"></a>
</h2>
<p>The proteomics test dataset was generated by culturing
<em>Pseudomonas putida</em> KT2440 in 500-mL shaken flasks filled with
50 mL media at 30°C and 250 rpm. The media were either synthetic de Bont
medium supplemented with 20 mM glucose or 30 mM acetate, or LB medium.
Samples were taken in the mid-exponential phase, and cells were
harvested by centrifugation at 17,000 x g for 2 min at 4°C. After
removal of the supernatant, cell pellets were frozen and kept at −80°C
until proteomics analysis was performed as described previously <span class="citation">(<a href="#ref-rennig2019">Rennig <em>et al.</em>,
2019</a>)</span>. No normalization of protein abundances was performed
prior to this analysis.</p>
<p>The path to the example dataset can be extracted with:</p>
<div class="sourceCode" id="cb3"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/base/system.file.html" class="external-link">system.file</a></span><span class="op">(</span><span class="st">"prot_KT2440_glc_ac_LB.txt"</span>, package <span class="op">=</span> <span class="st">"VisomX"</span><span class="op">)</span></span></code></pre></div>
<p>[1]
“/home/runner/work/_temp/Library/VisomX/prot_KT2440_glc_ac_LB.txt”</p>
<div class="section level3">
<h3 id="data-format">Data Format<a class="anchor" aria-label="anchor" href="#data-format"></a>
</h3>
<p><code>VisomX</code> requires a tabular input, either in the form of
an R dataframe, or stored in an <em>XLS</em>, <em>XLSX</em>,
<em>CSV</em>, <em>TSV</em>, or <em>TXT</em> file. Three types of columns
are required:</p>
<ol style="list-style-type: decimal">
<li><p>A column containing protein identifiers. These can be Uniprot
identifiers, Ensembl Gene IDs, Entrez Gene IDs, etc.</p></li>
<li><p>A column containing names. Gene Symbols are commonly used for
this purpose.</p></li>
<li><p>Columns containing protein abundances of the different samples in
the experiment. These columns should have a common prefix (e.g.,
<em>abundance.</em>) used for their identification within the table.
Furthermore, replicates are indicated by an underscore followed by a
number. For example, <em>abundance.KT2440_Glc_1</em>,
<em>abundance.KT2440_Glc_2</em>, <em>abundance.KT2440_Glc_3</em>, and
<em>abundance.KT2440_Glc_4</em> indicate four replicates for the
condition <em>KT2440_Glc</em>.</p></li>
<li><p><em>Optional</em>: Columns containing <em>+</em> and <em>-</em>
indicating contaminant proteins and decoy database hits that should be
filtered out.</p></li>
</ol>
</div>
</div>
<div class="section level2">
<h2 id="run-a-complete-proteomics-analysis-workflow">Run a Complete Proteomics Analysis Workflow<a class="anchor" aria-label="anchor" href="#run-a-complete-proteomics-analysis-workflow"></a>
</h2>
<div class="section level3">
<h3 id="read-data">Read Data<a class="anchor" aria-label="anchor" href="#read-data"></a>
</h3>
<p>The function <code><a href="../reference/prot.read_data.html">prot.read_data()</a></code> accepts either an R
dataframe or the path to a table file with extension ‘.xlsx’, ‘.xls’,
‘.csv’, ‘.tsv’, or ‘.txt’. It creates a
<code>SummarizedExperiment</code> object and performs feature filtering
based on indicated contaminants or unspecific identification, a defined
relative standard deviation threshold, or the presence of missing
values. For details about this function, run
<code><a href="../reference/prot.read_data.html">help("prot.read_data", "VisomX")</a></code>. For instructions on how
to work with <code>SummarizedExperiment</code> objects, see the <a href="https://www.bioconductor.org/help/course-materials/2019/BSS2019/04_Practical_CoreApproachesInBioconductor.html" class="external-link">Bioconductor
manual</a> Here, we filter proteins that are not present in at least 6
out of 9 samples.</p>
<div class="sourceCode" id="cb4"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Read data, incl. filtering</span></span>
<span><span class="va">data</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/prot.read_data.html">prot.read_data</a></span><span class="op">(</span></span>
<span>          data <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/system.file.html" class="external-link">system.file</a></span><span class="op">(</span><span class="st">"prot_KT2440_glc_ac_LB.txt"</span>, package <span class="op">=</span> <span class="st">"VisomX"</span><span class="op">)</span>,</span>
<span>          pfx <span class="op">=</span> <span class="st">"abundance."</span>,</span>
<span>          id <span class="op">=</span> <span class="st">"Ensembl Gene ID"</span>,</span>
<span>          name <span class="op">=</span> <span class="st">"Gene Symbol"</span>,</span>
<span>          filt_type <span class="op">=</span> <span class="st">"fraction"</span>,</span>
<span>          filt_min <span class="op">=</span> <span class="fl">0.66</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## Removing proteins with no defined name in column "Gene Symbol" or "Ensembl Gene ID". (Removed: 5 proteins)</span></span></code></pre>
<pre><code><span><span class="co">## Writing experimental design file to: /home/runner/work/VisomX/VisomX/vignettes/experimental_design.txt</span></span></code></pre>
<pre><code><span><span class="co">## Generating SummarizedExperiment.</span></span></code></pre>
<pre><code><span><span class="co">## 490 out of 3067 proteins were removed from the dataset due to missing values.</span></span>
<span><span class="co">## </span></span>
<span><span class="co">## Identified conditions:</span></span>
<span><span class="co">##  Glc, Ac, LB</span></span>
<span><span class="co">## No samples have all proteins as NA.</span></span>
<span><span class="co">## No samples to remove.</span></span></code></pre>
</div>
<div class="section level3">
<h3 id="run-workflow">Run Workflow<a class="anchor" aria-label="anchor" href="#run-workflow"></a>
</h3>
<p>The wrapper function <code><a href="../reference/prot.workflow.html">prot.workflow()</a></code> runs a complete
analysis workflow and exports the results as tabular TXT files as well
as (optionally) a report in PDF and HTML format. The following settings
usually need closer consideration:</p>
<ul>
<li><p>We use the <em>“SampMin”</em> method for missing value imputation
(<code>imp_fun</code>), which has been shown to be among the most
reliable methods for both missing-not-at-random (MNAR) and
missing-completely-at-random types of missing values <span class="citation">(<a href="#ref-liu2021">Liu &amp; Dongre,
2021</a>)</span>.</p></li>
<li><p>For the types of comparisons to perform (<code>type</code>), we
choose <em>“all”</em> to cover all possible contrasts. Alternatively,
one can compare every condition with a single control condition by
choosing <code>type = 'control'</code> and defining the control
condition with <code>control</code>. If only very specific contrasts are
of interest, <code>type = 'manual'</code> can be chosen and the desired
contrasts can be defined as a vector of strings in the form
<code>contrast = c('ConditionA_vs_ConditionB', ConditionC_vs_ConditionD')</code>.
The <code>_vs_</code> substring is hereby essential!</p></li>
<li><p>We set the significance threshold for adjusted p values
(<code>alpha</code>) at 0.05. <code>VisomX</code> tests for differential
expression of proteins based on protein-wise linear models and<br>
empirical Bayes statistics using <code>limma</code>. False Discovery
Rates are estimated using <code>fdrtool</code>.</p></li>
<li><p>We define 1 as the <em>relevance threshold</em> for log2 fold
changes (<code>lfc</code>), corresponding to fold changes of 2 and 0.5,
to be considered significant.</p></li>
<li><p>We choose <code>pathway_enrichment = TRUE</code> to perform
pathway enrichment analysis. We test for enriched KEGG pathways by
setting <code>pathway_kegg = TRUE</code> and defining a
<code>kegg_organism</code>. We also test for enriched pathways extracted
from the BioCyc database as a tab-delimited TXT file included in
<code>VisomX</code>. To test against a set of pathways other than those
listed in KEGG, we provide a dataframe object with columns
<em>‘Pathway’</em> and <em>‘Accession’</em>. <strong>Important</strong>:
The names or IDs used for proteins/genes listed under ‘Accession’ need
to correspond to the IDs initially used to create the
SummarizedExperiment with <code><a href="../reference/prot.read_data.html">prot.read_data()</a></code>.</p></li>
<li><p>Lastly, we choose whether a report shall be generated and
exported into a subfolder of the current working directory (omitted
because of the long run time associated).</p></li>
</ul>
<div class="sourceCode" id="cb9"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Read TXT file with BioCyc pathways for P. putida KT2440</span></span>
<span><span class="va">custom_df</span> <span class="op">&lt;-</span></span>
<span>  <span class="fu"><a href="https://rdrr.io/r/utils/read.table.html" class="external-link">read.table</a></span><span class="op">(</span></span>
<span>    <span class="fu"><a href="https://rdrr.io/r/base/system.file.html" class="external-link">system.file</a></span><span class="op">(</span><span class="st">"BioCyc_pathways_KT2440.txt"</span>, package <span class="op">=</span> <span class="st">"VisomX"</span><span class="op">)</span>,</span>
<span>    sep <span class="op">=</span> <span class="st">"\t"</span>,</span>
<span>    header <span class="op">=</span> <span class="cn">T</span>,</span>
<span>    stringsAsFactors <span class="op">=</span> <span class="cn">F</span>,</span>
<span>    fill <span class="op">=</span> <span class="cn">T</span>,</span>
<span>    na.strings <span class="op">=</span> <span class="st">""</span>,</span>
<span>    quote <span class="op">=</span> <span class="st">""</span>,</span>
<span>    comment.char <span class="op">=</span> <span class="st">""</span>,</span>
<span>    check.names <span class="op">=</span> <span class="cn">F</span></span>
<span>  <span class="op">)</span></span>
<span></span>
<span><span class="co"># Run workflow and export results</span></span>
<span><span class="va">results</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/prot.workflow.html">prot.workflow</a></span><span class="op">(</span>se <span class="op">=</span> <span class="va">data</span>,</span>
<span>        imp_fun <span class="op">=</span> <span class="st">"SampMin"</span>,</span>
<span>        type <span class="op">=</span> <span class="st">"all"</span>,</span>
<span>        alpha <span class="op">=</span> <span class="fl">0.05</span>,</span>
<span>        lfc <span class="op">=</span> <span class="fl">1</span>,</span>
<span>        pathway_enrichment <span class="op">=</span> <span class="cn">TRUE</span>,</span>
<span>        pathway_kegg <span class="op">=</span> <span class="cn">TRUE</span>,</span>
<span>        kegg_organism <span class="op">=</span> <span class="st">"ppu"</span>,</span>
<span>        custom_pathways <span class="op">=</span> <span class="va">custom_df</span>,</span>
<span>        volcano.add_names <span class="op">=</span> <span class="cn">FALSE</span>, <span class="co"># show protein labels in volcano</span></span>
<span>        report <span class="op">=</span> <span class="cn">FALSE</span>,</span>
<span>        out.dir <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/tempfile.html" class="external-link">tempdir</a></span><span class="op">(</span><span class="op">)</span></span>
<span>        <span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## Running proteomics workflow in the directory:/tmp/Rtmp5T4KSJ</span></span></code></pre>
<p>Tested contrasts: Glc_vs_Ac, Glc_vs_LB, Ac_vs_LB</p>
<pre><code><span><span class="co">## 316 proteins were found to be differentially expressed with alpha = 0.05 and |log2(fold change)| &gt; 1.</span></span></code></pre>
<pre><code><span><span class="co">## Save results as tab-delimited table to: /tmp/Rtmp5T4KSJ/results.txt</span></span></code></pre>
<pre><code><span><span class="co">## Save RData object</span></span></code></pre>
<pre><code><span><span class="co">## Writing results of KEGG pathway enrichment analysis to: /tmp/Rtmp5T4KSJ/pora_kegg_contrast...txt'</span></span></code></pre>
<pre><code><span><span class="co">## Writing results of custom pathway enrichment analysis to: /tmp/Rtmp5T4KSJ/pora_custom_contrast...txt</span></span></code></pre>
</div>
</div>
<div class="section level2">
<h2 id="custom-workflows">Custom workflows<a class="anchor" aria-label="anchor" href="#custom-workflows"></a>
</h2>
<p>If further customization of certain computational steps are required,
all steps included within <code><a href="../reference/prot.workflow.html">prot.workflow()</a></code> can also be
performed manually. The following steps are required to end up with the
same results list object as obtained with the workflow function:</p>
<div class="sourceCode" id="cb16"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># 1.  Perform data normalization: </span></span>
<span><span class="va">prot_norm</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/prot.normalize_vsn.html">prot.normalize_vsn</a></span><span class="op">(</span><span class="va">se</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># 2.  Missing value imputation</span></span>
<span></span>
<span><span class="co">#    a.   with imp_fun == "MinProb":</span></span>
<span>        <span class="va">prot_imp</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/prot.impute.html">prot.impute</a></span><span class="op">(</span><span class="va">prot_norm</span>, fun <span class="op">=</span> <span class="va">imp_fun</span>, q <span class="op">=</span> <span class="va">q</span><span class="op">)</span></span>
<span></span>
<span><span class="co">#    b.  with imp_fun == "knn":</span></span>
<span>        <span class="va">prot_imp</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/prot.impute.html">prot.impute</a></span><span class="op">(</span><span class="va">prot_norm</span>, fun <span class="op">=</span> <span class="va">imp_fun</span>, rowmax <span class="op">=</span> <span class="va">knn.rowmax</span><span class="op">)</span></span>
<span></span>
<span><span class="co">#    c.  for all other methods:</span></span>
<span>        <span class="va">prot_imp</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/prot.impute.html">prot.impute</a></span><span class="op">(</span><span class="va">prot_norm</span>, fun <span class="op">=</span> <span class="va">imp_fun</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># 3.  Principal component analysis</span></span>
<span><span class="va">prot_pca</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/prot.pca.html">prot.pca</a></span><span class="op">(</span><span class="fu">SummarizedExperiment</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/SummarizedExperiment/man/SummarizedExperiment-class.html" class="external-link">assay</a></span><span class="op">(</span><span class="va">prot_imp</span><span class="op">)</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># 4.  Test for differential expression by empirical Bayes moderation of a linear model and defined contrasts:</span></span>
<span><span class="va">prot_diff</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/prot.test_diff.html">prot.test_diff</a></span><span class="op">(</span><span class="va">prot_imp</span>, type <span class="op">=</span> <span class="va">type</span>, control <span class="op">=</span> <span class="va">control</span>, test <span class="op">=</span> <span class="va">contrast</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># 5.  Denote significantly differential proteins: </span></span>
<span><span class="va">prot_dep</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/prot.add_rejections.html">prot.add_rejections</a></span><span class="op">(</span><span class="va">prot_diff</span>, alpha <span class="op">=</span> <span class="va">alpha</span>, lfc <span class="op">=</span> <span class="va">lfc</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># 6.  Generate a results table: </span></span>
<span><span class="va">prot_res</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/prot.get_results.html">prot.get_results</a></span><span class="op">(</span><span class="va">prot_dep</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># 7.  Pathway enrichment analysis</span></span>
<span></span>
<span><span class="co">#    a. Get vector of tested contrasts: </span></span>
<span>        <span class="va">contrasts</span> <span class="op">&lt;-</span> <span class="fu">SummarizedExperiment</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/SummarizedExperiment/man/SummarizedExperiment-class.html" class="external-link">rowData</a></span><span class="op">(</span><span class="va">prot_dep</span><span class="op">)</span> <span class="op">%&gt;%</span></span>
<span>          <span class="fu"><a href="https://rdrr.io/r/base/data.frame.html" class="external-link">data.frame</a></span><span class="op">(</span>check.names <span class="op">=</span> <span class="cn">FALSE</span><span class="op">)</span> <span class="op">%&gt;%</span></span>
<span>          <span class="fu">select</span><span class="op">(</span><span class="fu">ends_with</span><span class="op">(</span><span class="st">"_diff"</span><span class="op">)</span><span class="op">)</span> <span class="op">%&gt;%</span></span>
<span>          <span class="fu"><a href="https://rdrr.io/r/base/colnames.html" class="external-link">colnames</a></span><span class="op">(</span><span class="op">)</span> <span class="op">%&gt;%</span> <span class="fu">str_replace_all</span><span class="op">(</span><span class="st">"_diff"</span>, <span class="st">""</span><span class="op">)</span></span>
<span>        </span>
<span><span class="co">#    b. Perform pathway enrichment analysis</span></span>
<span>        <span class="va">res.pathway</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/enrich_pathways.html">enrich_pathways</a></span><span class="op">(</span><span class="va">prot_dep</span>, <span class="va">contrasts</span>, alpha_pathways <span class="op">=</span> <span class="va">alpha_pathways</span>,</span>
<span>                                   pathway_kegg <span class="op">=</span> <span class="va">pathway_kegg</span>, kegg_organism <span class="op">=</span> <span class="va">kegg_organism</span>,</span>
<span>                                   custom_pathways <span class="op">=</span> <span class="va">custom_pathways</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># 8.  Assemble results into list</span></span>
<span>        <span class="va">results</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/list.html" class="external-link">list</a></span><span class="op">(</span>data <span class="op">=</span> <span class="fu">SummarizedExperiment</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/SummarizedExperiment/man/SummarizedExperiment-class.html" class="external-link">rowData</a></span><span class="op">(</span><span class="va">se</span><span class="op">)</span>, se <span class="op">=</span> <span class="va">se</span>, norm <span class="op">=</span> <span class="va">prot_norm</span>,</span>
<span>                  imputed <span class="op">=</span> <span class="va">prot_imp</span>, pca <span class="op">=</span> <span class="va">prot_pca</span>, diff <span class="op">=</span> <span class="va">prot_diff</span>, dep <span class="op">=</span> <span class="va">prot_dep</span>,</span>
<span>                  results <span class="op">=</span> <span class="va">results</span>, param <span class="op">=</span> <span class="va">param</span><span class="op">)</span></span></code></pre></div>
</div>
<div class="section level2">
<h2 id="plot-results">Plot Results<a class="anchor" aria-label="anchor" href="#plot-results"></a>
</h2>
<p>Most plots available in <code>VisomX</code> are generated
automatically within the report. Additionally, the figures are
automatically exported as PNG and PDF files if argument
<code>export = TRUE</code> when running
<code><a href="../reference/prot.workflow.html">prot.workflow()</a></code>.</p>
<p>To customize specific plot or generate them post-hoc, this section
provides an overview of available plotting functions.</p>
<div class="section level3">
<h3 id="protein-coverage">Protein Coverage<a class="anchor" aria-label="anchor" href="#protein-coverage"></a>
</h3>
<p>Protein identifications per sample:</p>
<p><img src="vignette_proteomics_files/figure-html/numbers-1.png" width="432" style="display: block; margin: auto;"></p>
<p>Gene coverage in all samples:</p>
<p><img src="vignette_proteomics_files/figure-html/coverage-1.png" width="576" style="display: block; margin: auto;"></p>
</div>
<div class="section level3">
<h3 id="effect-of-normalization">Effect of Normalization<a class="anchor" aria-label="anchor" href="#effect-of-normalization"></a>
</h3>
<p>The purpose of the following graphs is to determine if there is a
systematic trend in the standard deviation of the data as a function of
overall expression. These graphs are based on the assumption that most
proteins are not differentially expressed, so the running median is a
reasonable estimate of the standard deviation of the data at the feature
level, as a function of the mean. After vsn normalization, the running
median should be approximately a horizontal line. While a completely
flat curve of the square root of variance over the mean may seem like
the goal of such transformations, this may be unreasonable in the case
of datasets with many true differences due to the experimental
conditions. There may be random fluctuations, but there should not be a
general trend. If this is not the case, it usually indicates a data
quality problem or is the result of inadequate data preprocessing.</p>
<p>SD-Rank(mean) plot before normalization:</p>
<p><img src="vignette_proteomics_files/figure-html/meanSDplot1-1.png" width="480" style="display: block; margin: auto;"></p>
<p>SD-Rank(mean) plot after normalization:</p>
<p><img src="vignette_proteomics_files/figure-html/meanSDplot2-1.png" width="480" style="display: block; margin: auto;"></p>
<p>Box plots of the distributions before and after normalization:</p>
<p><img src="vignette_proteomics_files/figure-html/norm-1.png" width="700" style="display: block; margin: auto;"></p>
</div>
<div class="section level3">
<h3 id="missing-values">Missing Values<a class="anchor" aria-label="anchor" href="#missing-values"></a>
</h3>
<p>Missing values in the data set must be imputed. The downstream
analysis can be strongly influenced by the imputation. Therefore, the
selection of an appropriate method is crucial. The data can be missing
at random (MAR), for example if proteins are quantified in some
replicates but not in others, without any bias towards low intensities.
This class of missing values can arise when the peptide sequence is
mapped incorrectly or software erroneously assigns shared peptides to
precursors leading to misidentification in some samples and missing
values in others.<br>
MNAR may result from experimental effects such as (1) enzyme
miscleavages, (2) true presence/absence in the biological samples and
(3) instrumentation effects (when peptide measurements are low in
abundance compared to background noise or constitute low ionization
efficiency). Because values are missing because of low abundant nature
of the respective proteins, this category of missing values is
considered left-censored, i.e., the distribution of values (if present
in the data) would fall on the left tail of the total observations in
the dataset. A detailed treatise about missing values in the context of
proteomics data can be found here: <a href="http://www.bioconductor.org/packages/release/bioc/vignettes/DEP/inst/doc/MissingValues.html" class="external-link">Missing
value handling</a></p>
<div class="section level4">
<h4 id="define-missing-value-type">Define Missing Value Type<a class="anchor" aria-label="anchor" href="#define-missing-value-type"></a>
</h4>
<p>To asses the type of missing data (random or not), the following heat
map of missing values and density distributions of proteins with/without
missing values provide a visual aid.<br>
If data is randomly missing, use the k-nearest neighbor (“<em>knn</em>”)
or maximum likelihood (“<em>MLE</em>”) options for imputation. If the
missing data is biased to certain samples (e.g. controls) which are
expected to be depleted of certain proteins and/or if there is a clear
bias of missing values towards low abundances, use the “<em>QRILC</em>”,
“<em>MinProb</em>”, “<em>SampMin</em>”, or “<em>man</em>” options for
imputation. For a more detailed description of different imputation
methods, see the <a href="https://bioconductor.org/packages/3.15/bioc/html/MSnbase.html" class="external-link"><em>MSnbase</em></a>
vignette and in particular the <em>impute</em> function description. The
“<strong>SampMin</strong>” method was proposed by <a href="https://doi.org/10.1093/bib/bbaa112" class="external-link">Liu and Dongre (2020)</a>)
and shown to outperform other common methods for MNAR, left-censored
datasets.</p>
<p><img src="vignette_proteomics_files/figure-html/missval-1.png" width="672" style="display: block; margin: auto;"></p>
<p>To check whether missing values are biased toward low-intensity
proteins (i.e., ‘left-censored’), densities and cumulative proportions
are plotted for proteins with and without missing values.</p>
<p><img src="vignette_proteomics_files/figure-html/missval2-1.png" width="672" style="display: block; margin: auto;"></p>
</div>
<div class="section level4">
<h4 id="effects-of-imputation">Effects of Imputation<a class="anchor" aria-label="anchor" href="#effects-of-imputation"></a>
</h4>
<p>The effects of data normalization and imputation on the distributions
are visualized via density distributions:</p>
<p><img src="vignette_proteomics_files/figure-html/density-1.png" width="672" style="display: block; margin: auto;"></p>
</div>
</div>
<div class="section level3">
<h3 id="quality-control">Quality Control<a class="anchor" aria-label="anchor" href="#quality-control"></a>
</h3>
<div class="section level4">
<h4 id="principal-component-analysis">Principal Component Analysis<a class="anchor" aria-label="anchor" href="#principal-component-analysis"></a>
</h4>
<p>To get a high-level overview of the data, the unsupervised method
<em>principal component analysis (PCA)</em> reduces the data
dimensionality (i.e., number of proteins included in the analysis) while
retaining most of the data information. PCA is an unsupervised method
for reducing the dimensionality of high-dimensional data. The goal of
PCA is to find a new set of variables, called “principal components,”
that capture most of the data information while reducing the number of
dimensions. This is achieved by finding a new coordinate system that
aligns with the directions of maximum variance in the data. The first
principal component captures the most variation in the data, the second
component captures the second-most variation, and so on. By retaining
only the first few principal components, the data can be reduced to a
lower-dimensional representation while retaining most of the important
information. This can be useful for visualizing the data, reducing
noise, or improving the performance of machine learning algorithms.</p>
<div class="section level5">
<h5 id="scree-plot">Scree Plot<a class="anchor" aria-label="anchor" href="#scree-plot"></a>
</h5>
<p>A scree plot in principal component analysis (PCA) is a graphical
representation of the eigenvalues of the principal components. It helps
to determine the number of significant components in a PCA analysis by
plotting the eigenvalues against the component number. The “scree”
refers to the downward-sloping “elbow” shape that often appears in these
plots, which is used as a visual aid for determining the number of
important components. The idea is to keep components with large
eigenvalues and discard the others, as they are likely to be noise or
have little contribution to the data variance.</p>
<p><img src="vignette_proteomics_files/figure-html/screeplot-1.png" width="672" style="display: block; margin: auto;"></p>
<p>From the Scree Plot, we see that the first two PCs explain about
two-thirds of the observed variation in the dataset. Further components
provide little explanatory value.</p>
</div>
<div class="section level5">
<h5 id="d-pca-plots">2D PCA plots<a class="anchor" aria-label="anchor" href="#d-pca-plots"></a>
</h5>
<p>A 2D PCA plot is a visualization of a dataset after undergoing
principal component analysis (PCA) and reducing the data to two
dimensions. In a 2D PCA plot, each data point is represented as a dot on
a 2D graph, with the x-axis and y-axis representing the first and second
principal components, respectively. These new coordinates, or principal
components (PCs), are not real variables generated by the system and
therefore, applying PCA to a dataset loses its interpretability.</p>
<p>However, despite the loss of interpretability, PCA can be very useful
in identifying batch effects and assessing which original samples are
similar and different from each other. By plotting the data in the 2D
PCA space, it can be easier to visualize and identify clusters of
similar samples, and distinguish samples that are different from each
other. This information can then be used to better understand the
underlying structure of the data, and improve the accuracy of downstream
analyses and applications.</p>
<p>The function <code>prot.plot_pca</code> internally performs PCA
analysis and requires a SummarizedExperiment object. Thus, the
<code>imp</code> rather than the <code>pca</code> object is used as
input.</p>
<p><img src="vignette_proteomics_files/figure-html/plot_pca_1_2-1.png" width="672" style="display: block; margin: auto;"></p>
<p><img src="vignette_proteomics_files/figure-html/plot_pca_1_3-1.png" width="672" style="display: block; margin: auto;"></p>
<p>We can see that the three conditions cluster well in PC1, less in
PC2, and not at all in PC3 which explains only 9.5% variation within the
dataset.</p>
</div>
<div class="section level5">
<h5 id="loadings-plot">Loadings plot<a class="anchor" aria-label="anchor" href="#loadings-plot"></a>
</h5>
<p>PCA loadings are the coefficients of the linear combination of the
original variables (proteins) from which the principal components (PCs)
are generated. They describe how much each gene contributes to a
particular principal component. Large loadings (positive or negative)
indicate that a particular variable has a strong relationship to a
particular principal component. The sign of a loading indicates whether
a variable and a principal component are positively or negatively
correlated.</p>
<p>The <code><a href="../reference/prot.plot_loadings.html">prot.plot_loadings()</a></code> function accept a
<code>pca</code> object generated with <code><a href="../reference/prot.pca.html">prot.pca()</a></code>.</p>
<p><img src="vignette_proteomics_files/figure-html/loadingsPlot-1.png" width="672" style="display: block; margin: auto;"></p>
<p>High loadings on PCs that explain the observed variations to a high
degree are a direct indicator that the respective proteins play an
important role in distinguishing the tested conditions. In PC1, the
<em>LB</em> group clustered at a great distance to the <em>Glc</em> and
<em>Ac</em> groups. The high loading of <strong>BkdA1</strong> on PC1
suggests that differences in the abundance of this protein are a
significant factor distinguishing the conditions. This makes sense,
since it is a subunit of the branched-chain alpha-keto acid
dehydrogenase complex, involved in the degradation of branched chain
amino acid, an abundand carbon source in LB medium that is absent in the
other conditions.</p>
</div>
</div>
<div class="section level4">
<h4 id="correlation-matrix">Correlation Matrix<a class="anchor" aria-label="anchor" href="#correlation-matrix"></a>
</h4>
<p>A correlation matrix is a type of data matrix that shows the pairwise
relationships between variables in a dataset. In quality control, a
correlation matrix can be used to assess the quality of the data by
examining the relationships between samples. By plotting the samples in
a correlation matrix, it is possible to identify groups of samples that
are highly similar to each other, and to differentiate samples that are
not similar. This can be useful for identifying and correcting batch
effects, which can affect the accuracy of the data analysis. For
example, if some samples are highly correlated within a batch, but not
correlated with samples from another batch, this may indicate that the
samples from different batches have been processed differently, and that
further adjustments or normalization may be necessary to ensure accurate
analysis. A correlation matrix is usually represented as a heatmap,
where each cell in the matrix represents the pairwise correlation
between two samples. Samples that are highly correlated will appear in
similar colors in the heatmap, while samples that are not correlated
will appear in different colors.</p>
<p>The Pearson coefficient, also known as Pearson’s correlation
coefficient, is a statistic that measures the strength and direction of
the linear relationship between two continuous variables.</p>
<p>The Pearson coefficient ranges from -1 to 1, with a value of 1
indicating a perfect positive linear relationship, a value of -1
indicating a perfect negative linear relationship, and a value of 0
indicating no linear relationship. Positive values of the Pearson
coefficient indicate that as one variable increases, the other variable
also increases. Negative values of the Pearson coefficient indicate that
as one variable increases, the other variable decreases.</p>
<div class="section level5">
<h5 id="correlation-matrix-for-samples-with-all-proteins">Correlation matrix for samples with all proteins<a class="anchor" aria-label="anchor" href="#correlation-matrix-for-samples-with-all-proteins"></a>
</h5>
<p>As most proteins within a proteomics dataset usually have little
variation, the pearson coefficients observed are usually high. Thus, we
set a lower threshold for the applied color scale to 0.8.</p>
<p><img src="vignette_proteomics_files/figure-html/correlation_heatmap_all-1.png" width="672" style="display: block; margin: auto;"></p>
</div>
</div>
</div>
<div class="section level3">
<h3 id="differential-expression-analysis">Differential expression analysis<a class="anchor" aria-label="anchor" href="#differential-expression-analysis"></a>
</h3>
<p>The following plots are a visual representation of differential
expression (DE) analysis. DE analysis is the process of comparing gene
or protein expression levels between two or more groups of samples in
order to identify genes or proteins that are differentially expressed
between the groups. The aim of DE analysis is to identify genes or
proteins that show statistically significant differences in expression
levels between the different groups.</p>
<div class="section level4">
<h4 id="correlation-matrix-with-significant-proteins">Correlation Matrix with Significant Proteins<a class="anchor" aria-label="anchor" href="#correlation-matrix-with-significant-proteins"></a>
</h4>
<p>In the following correlation plot, we consider only significant
proteins:</p>
<p><img src="vignette_proteomics_files/figure-html/correlation_heatmap-1.png" width="672" style="display: block; margin: auto;"></p>
</div>
<div class="section level4">
<h4 id="heatmaps">Heatmaps<a class="anchor" aria-label="anchor" href="#heatmaps"></a>
</h4>
<p>Heat maps are either used to visualize the expression level of each
protein in each sample or the log2 fold changes for each tested
contrast. The heat maps give an overview of all proteins with
significantly different abundances (rows) in all samples (bars). This
allows general trends to be identified, for example when one sample or
replicate differs from others. In addition, clustering of samples
(columns) can point to more similar samples and clustering of proteins
(rows) can indicate proteins that behave in a similar way. By
considering only significant proteins we can get a better overview of
the expression levels of proteins that are relevant to the comparison of
the conditions.</p>
<p><img src="vignette_proteomics_files/figure-html/heatmap_1-1.png" width="672" style="display: block; margin: auto;"></p>
<p><img src="vignette_proteomics_files/figure-html/heatmap_2-1.png" width="806.4" style="display: block; margin: auto;"></p>
</div>
<div class="section level4">
<h4 id="volcano-plots">Volcano plot(s)<a class="anchor" aria-label="anchor" href="#volcano-plots"></a>
</h4>
<p>Volcano plots can be used to visualize a particular contrast
(comparison between two samples). This allows to examine the gene
enrichment between two samples (x-axis) and their corresponding
(adjusted) p-values (y-axis).</p>
<p><img src="vignette_proteomics_files/figure-html/volcano-1.png" width="768" style="display: block; margin: auto;"></p>
</div>
</div>
<div class="section level3">
<h3 id="pathway-overrepresentation-analysis">Pathway overrepresentation analysis<a class="anchor" aria-label="anchor" href="#pathway-overrepresentation-analysis"></a>
</h3>
<p>Pathway overrepresentation analysis is a method used to identify
pathways that are significantly enriched in a set of genes or proteins.
It is a useful tool for interpreting the results of gene or protein
expression studies, as it can identify pathways that are likely to be
involved in the studied biological process.</p>
<p>We use <strong><em>lollipop charts</em></strong> to visualize
significantly enriched pathways. In these plots, each enriched pathway
is represented as a horizontal line, with the length indicating what
fraction of the total genes constituting the pathway were found to be
enriched (e.g, if a pathway consists of 10 genes of which 6 were found
to be overrepresented, the <em>Rich factor</em> is 0.6). The circle size
at the end of the ‘lollipops’ indicates the absolute number of enriched
genes in the pathway, while its color indicates the adjusted p value
associated with the finding.</p>
<p>Instead of Venn diagrams to illustrate the relationship between
different gene sets and how they overlap, we use <strong>UpSet
plots</strong> [<span class="citation">Lex <em>et al.</em> (<a href="#ref-lex2014">2014</a>)</span>]<span class="citation">(<a href="#ref-conway2017">Conway <em>et al.</em>, 2017</a>)</span>. In a
Venn diagram, the intersections between sets are represented as
overlapping circles, with the area of the overlap representing the
number of elements in the intersection. However, Venn diagrams become
less effective as the number of sets increases, since the overlapping
circles can become complex and difficult to interpret. In contrast,
UpSet plots are more effective for larger numbers of sets, as they
represent the intersection between sets as bars across the x-axis, with
the height of the bars representing the number of elements in the
intersection. The matrix below the bar chart provides details about the
set intersections. This makes it easier to compare and visualize the
relationships between sets, even when there are many sets. Further
instructions on how to interpret UpSet plots can be found at <a href="https://upset.app/" class="external-link uri">https://upset.app/</a>.</p>
<div class="section level4">
<h4 id="kegg-pathways">KEGG pathways<a class="anchor" aria-label="anchor" href="#kegg-pathways"></a>
</h4>
<p>In the following plot, we have used the KEGG pathways to visualize
the significant proteins that are differentially expressed within the
tested contrast.</p>
<p>We can generate the plot showing the overrepresented KEGG pathways
for a specific contrast:</p>
<p><img src="vignette_proteomics_files/figure-html/enrichment_kegg-1.png" width="1056"><img src="vignette_proteomics_files/figure-html/enrichment_kegg-2.png" width="1056"></p>
<p>Further details about the enriched pathways can be extracted via
converting the <code>enrichResult</code> objects into a dataframe:</p>
<div class="sourceCode" id="cb17"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">df</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/as.data.frame.html" class="external-link">as.data.frame</a></span><span class="op">(</span><span class="va">results</span><span class="op">$</span><span class="va">pora_kegg_up</span><span class="op">[[</span><span class="st">"Glc_vs_Ac"</span><span class="op">]</span><span class="op">]</span><span class="op">)</span></span>
<span><span class="co"># Inspect the structure of the dataframe</span></span>
<span><span class="fu">glimpse</span><span class="op">(</span><span class="va">df</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## Rows: 10</span></span>
<span><span class="co">## Columns: 13</span></span>
<span><span class="co">## $ ID             <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span> "ppu00030"<span style="color: #949494;">, </span>"ppu01120"<span style="color: #949494;">, </span>"ppu01200"<span style="color: #949494;">, </span>"ppu00620"<span style="color: #949494;">, </span>"ppu000…</span></span>
<span><span class="co">## $ Description    <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span> "Pentose phosphate pathway"<span style="color: #949494;">, </span>"Microbial metabolism in d…</span></span>
<span><span class="co">## $ GeneRatio      <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span> "10/39"<span style="color: #949494;">, </span>"24/39"<span style="color: #949494;">, </span>"16/39"<span style="color: #949494;">, </span>"9/39"<span style="color: #949494;">, </span>"7/39"<span style="color: #949494;">, </span>"7/39"<span style="color: #949494;">, </span>"4/3…</span></span>
<span><span class="co">## $ BgRatio        <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span> "34/1819"<span style="color: #949494;">, </span>"320/1819"<span style="color: #949494;">, </span>"135/1819"<span style="color: #949494;">, </span>"54/1819"<span style="color: #949494;">, </span>"30/1819"…</span></span>
<span><span class="co">## $ RichFactor     <span style="color: #949494; font-style: italic;">&lt;dbl&gt;</span> 0.29411765<span style="color: #949494;">, </span>0.07500000<span style="color: #949494;">, </span>0.11851852<span style="color: #949494;">, </span>0.16666667<span style="color: #949494;">, </span>0.23333…</span></span>
<span><span class="co">## $ FoldEnrichment <span style="color: #949494; font-style: italic;">&lt;dbl&gt;</span> 13.717949<span style="color: #949494;">, </span>3.498077<span style="color: #949494;">, </span>5.527825<span style="color: #949494;">, </span>7.773504<span style="color: #949494;">, </span>10.882906<span style="color: #949494;">, </span>8.5…</span></span>
<span><span class="co">## $ zScore         <span style="color: #949494; font-style: italic;">&lt;dbl&gt;</span> 11.077880<span style="color: #949494;">, </span>7.284491<span style="color: #949494;">, </span>8.091047<span style="color: #949494;">, </span>7.477513<span style="color: #949494;">, </span>8.077173<span style="color: #949494;">, </span>6.99…</span></span>
<span><span class="co">## $ pvalue         <span style="color: #949494; font-style: italic;">&lt;dbl&gt;</span> 5.488843e-10<span style="color: #949494;">, </span>7.007909e-10<span style="color: #949494;">, </span>3.082218e-09<span style="color: #949494;">, </span>9.678631e-07<span style="color: #949494;">,</span>…</span></span>
<span><span class="co">## $ p.adjust       <span style="color: #949494; font-style: italic;">&lt;dbl&gt;</span> 1.156305e-08<span style="color: #949494;">, </span>1.156305e-08<span style="color: #949494;">, </span>3.390440e-08<span style="color: #949494;">, </span>7.984871e-06<span style="color: #949494;">,</span>…</span></span>
<span><span class="co">## $ qvalue         <span style="color: #949494; font-style: italic;">&lt;dbl&gt;</span> 8.483258e-09<span style="color: #949494;">, </span>8.483258e-09<span style="color: #949494;">, </span>2.487404e-08<span style="color: #949494;">, </span>5.858119e-06<span style="color: #949494;">,</span>…</span></span>
<span><span class="co">## $ geneID         <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span> "PP_1024/PP_1010/PP_3416/PP_3376/PP_3378/PP_1023/PP_338…</span></span>
<span><span class="co">## $ Count          <span style="color: #949494; font-style: italic;">&lt;int&gt;</span> 10<span style="color: #949494;">, </span>24<span style="color: #949494;">, </span>16<span style="color: #949494;">, </span>9<span style="color: #949494;">, </span>7<span style="color: #949494;">, </span>7<span style="color: #949494;">, </span>4<span style="color: #949494;">, </span>2<span style="color: #949494;">, </span>3<span style="color: #949494;">, </span>2</span></span>
<span><span class="co">## $ richFactor     <span style="color: #949494; font-style: italic;">&lt;dbl&gt;</span> 0.29411765<span style="color: #949494;">, </span>0.07500000<span style="color: #949494;">, </span>0.11851852<span style="color: #949494;">, </span>0.16666667<span style="color: #949494;">, </span>0.23333…</span></span></code></pre>
<p>No KEGG pathways were found to be UNDERrepresented for the contrast
“Glc_vs_Ac”:</p>
<div class="sourceCode" id="cb19"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">df</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/as.data.frame.html" class="external-link">as.data.frame</a></span><span class="op">(</span><span class="va">results</span><span class="op">$</span><span class="va">pora_kegg_dn</span><span class="op">[[</span><span class="st">"Glc_vs_Ac"</span><span class="op">]</span><span class="op">]</span><span class="op">)</span></span>
<span><span class="co"># Inspect the structure of the dataframe</span></span>
<span><span class="fu">glimpse</span><span class="op">(</span><span class="va">df</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## Rows: 0</span></span>
<span><span class="co">## Columns: 13</span></span>
<span><span class="co">## $ ID             <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span> </span></span>
<span><span class="co">## $ Description    <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span> </span></span>
<span><span class="co">## $ GeneRatio      <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span> </span></span>
<span><span class="co">## $ BgRatio        <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span> </span></span>
<span><span class="co">## $ RichFactor     <span style="color: #949494; font-style: italic;">&lt;dbl&gt;</span> </span></span>
<span><span class="co">## $ FoldEnrichment <span style="color: #949494; font-style: italic;">&lt;dbl&gt;</span> </span></span>
<span><span class="co">## $ zScore         <span style="color: #949494; font-style: italic;">&lt;dbl&gt;</span> </span></span>
<span><span class="co">## $ pvalue         <span style="color: #949494; font-style: italic;">&lt;dbl&gt;</span> </span></span>
<span><span class="co">## $ p.adjust       <span style="color: #949494; font-style: italic;">&lt;dbl&gt;</span> </span></span>
<span><span class="co">## $ qvalue         <span style="color: #949494; font-style: italic;">&lt;dbl&gt;</span> </span></span>
<span><span class="co">## $ geneID         <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span> </span></span>
<span><span class="co">## $ Count          <span style="color: #949494; font-style: italic;">&lt;int&gt;</span> </span></span>
<span><span class="co">## $ richFactor     <span style="color: #949494; font-style: italic;">&lt;dbl&gt;</span></span></span></code></pre>
</div>
<div class="section level4">
<h4 id="custom-pathways">Custom pathways<a class="anchor" aria-label="anchor" href="#custom-pathways"></a>
</h4>
<p>In the following plot, we have used the BioCyc pathways to visualize
the significant proteins that are differentially expressed within the
tested contrast.</p>
<p><img src="vignette_proteomics_files/figure-html/enrichment_custom_up-1.png" width="1056"><img src="vignette_proteomics_files/figure-html/enrichment_custom_up-2.png" width="1056"></p>
<p><img src="vignette_proteomics_files/figure-html/enrichment_custom_dn-1.png" width="1056"><img src="vignette_proteomics_files/figure-html/enrichment_custom_dn-2.png" width="1056"></p>
</div>
</div>
<div class="section level3">
<h3 id="plot-abundances-and-log2-fold-changes-of-selected-proteins">Plot Abundances and Log2 Fold Changes of selected proteins<a class="anchor" aria-label="anchor" href="#plot-abundances-and-log2-fold-changes-of-selected-proteins"></a>
</h3>
<p>Often, a specific set of proteins is of particular interest for an
analysis. The <code><a href="../reference/prot.plot_bar.html">prot.plot_bar()</a></code> allows visualization of
protein abundances across different conditions or their log2-fold ratio
for selected contrasts in a column chart.</p>
<p>For example, let’s have a look at the proteins constituting the TCA
cycle. For this, we can store all KEGG pathways in a list object using
the <code>prot.get_kegg_pathways</code> and then extract the respective
genes to create the plot. Since <code>prot.get_kegg_pathways</code>
returns locus IDs and not gene names, we convert the IDs into names by
accessing the primary information-containing dataframe within a
SummarizedExperiment object with <code>rowData()</code> from package
<code>SummarizedExperiment</code>:</p>
<p><img src="vignette_proteomics_files/figure-html/PlotBar-1.png" width="1248"></p>
<p>Lastly, we plot the log2 fold change values of TCA cycle proteins for
one of the tested contrasts:</p>
<p><img src="vignette_proteomics_files/figure-html/PlotBarFC-1.png" width="1248">
Other <code>type</code> option of this function include
<code>reference</code>, in which the log2(abundance) of all shown
proteins are centered around a reference protein, and
<code>abundance</code>, which shows the absolute log(abundance)
values.</p>
</div>
</div>
<div class="section level2">
<h2 id="references">References<a class="anchor" aria-label="anchor" href="#references"></a>
</h2>
<div id="refs" class="references csl-bib-body hanging-indent" entry-spacing="0">
<div id="ref-conway2017" class="csl-entry">
Conway J.R., Lex A. &amp; Gehlenborg N. (2017). UpSetR: an R package for
the visualization of intersecting sets and their properties.
Bioinformatics (Oxford, England) 33 (18): 2938–2940. <a href="https://doi.org/10.1093/bioinformatics/btx364" class="external-link">https://doi.org/10.1093/bioinformatics/btx364</a>.
</div>
<div id="ref-lex2014" class="csl-entry">
Lex A., Gehlenborg N., Strobelt H., Vuillemot R. &amp; Pfister H.
(2014). UpSet: Visualization of Intersecting Sets. IEEE transactions on
visualization and computer graphics 20 (12): 1983–1992. <a href="https://doi.org/10.1109/TVCG.2014.2346248" class="external-link">https://doi.org/10.1109/TVCG.2014.2346248</a>.
</div>
<div id="ref-liu2021" class="csl-entry">
Liu M. &amp; Dongre A. (2021). Proper imputation of missing values in
proteomics datasets for differential expression analysis. Briefings in
bioinformatics 22 (3): bbaa112. <a href="https://doi.org/10.1093/bib/bbaa112" class="external-link">https://doi.org/10.1093/bib/bbaa112</a>.
</div>
<div id="ref-rennig2019" class="csl-entry">
Rennig M., Mundhada H., Wordofa G.G., Gerngross D., Wulff T., Worberg
A., Nielsen A.T. &amp; Nørholm M.H.H. (2019). Industrializing a
Bacterial Strain for l-Serine Production through Translation Initiation
Optimization. ACS synthetic biology 8 (10): 2347–2358. <a href="https://doi.org/10.1021/acssynbio.9b00169" class="external-link">https://doi.org/10.1021/acssynbio.9b00169</a>.
</div>
<div id="ref-zhang2018" class="csl-entry">
Zhang X., Smits A.H., Tilburg G.B. van, Ovaa H., Huber W. &amp;
Vermeulen M. (2018). Proteome-wide identification of ubiquitin
interactions using UbIA-MS. Nature Protocols 13 (3): 530–550. <a href="https://doi.org/10.1038/nprot.2017.147" class="external-link">https://doi.org/10.1038/nprot.2017.147</a>.
</div>
</div>
</div>
  </main><aside class="col-md-3"><nav id="toc" aria-label="Table of contents"><h2>On this page</h2>
    </nav></aside>
</div>



    <footer><div class="pkgdown-footer-left">
  <p>Developed by Nicolas T. Wirth.</p>
</div>

<div class="pkgdown-footer-right">
  <p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.1.2.</p>
</div>

    </footer>
</div>





  </body>
</html>
