<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<title>Reliable and extensive analysis of transcriptomics data with package `VisomX` • VisomX</title>
<script src="../deps/jquery-3.6.0/jquery-3.6.0.min.js"></script><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<link href="../deps/bootstrap-5.3.1/bootstrap.min.css" rel="stylesheet">
<script src="../deps/bootstrap-5.3.1/bootstrap.bundle.min.js"></script><link href="../deps/font-awesome-6.5.2/css/all.min.css" rel="stylesheet">
<link href="../deps/font-awesome-6.5.2/css/v4-shims.min.css" rel="stylesheet">
<script src="../deps/headroom-0.11.0/headroom.min.js"></script><script src="../deps/headroom-0.11.0/jQuery.headroom.min.js"></script><script src="../deps/bootstrap-toc-1.0.1/bootstrap-toc.min.js"></script><script src="../deps/clipboard.js-2.0.11/clipboard.min.js"></script><script src="../deps/search-1.0.0/autocomplete.jquery.min.js"></script><script src="../deps/search-1.0.0/fuse.min.js"></script><script src="../deps/search-1.0.0/mark.min.js"></script><!-- pkgdown --><script src="../pkgdown.js"></script><link href="../extra.css" rel="stylesheet">
<meta property="og:title" content="Reliable and extensive analysis of transcriptomics data with package `VisomX`">
</head>
<body>
    <a href="#main" class="visually-hidden-focusable">Skip to contents</a>


    <nav class="navbar navbar-expand-lg fixed-top bg-light" data-bs-theme="light" aria-label="Site navigation"><div class="container">

    <a class="navbar-brand me-2" href="../index.html">VisomX</a>

    <small class="nav-text text-muted me-auto" data-bs-toggle="tooltip" data-bs-placement="bottom" title="">0.0.94</small>


    <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbar" aria-controls="navbar" aria-expanded="false" aria-label="Toggle navigation">
      <span class="navbar-toggler-icon"></span>
    </button>

    <div id="navbar" class="collapse navbar-collapse ms-3">
      <ul class="navbar-nav me-auto">
<li class="nav-item"><a class="nav-link" href="../reference/index.html">Reference</a></li>
<li class="active nav-item dropdown">
  <button class="nav-link dropdown-toggle" type="button" id="dropdown-articles" data-bs-toggle="dropdown" aria-expanded="false" aria-haspopup="true">Articles</button>
  <ul class="dropdown-menu" aria-labelledby="dropdown-articles">
<li><a class="dropdown-item" href="../articles/vignette_proteomics.html">Reliable and extensive analysis of Proteomics data with package `VisomX`</a></li>
    <li><a class="dropdown-item" href="../articles/vignette_transcriptomics.html">Reliable and extensive analysis of transcriptomics data with package `VisomX`</a></li>
  </ul>
</li>
      </ul>
<ul class="navbar-nav">
<li class="nav-item"><form class="form-inline" role="search">
 <input class="form-control" type="search" name="search-input" id="search-input" autocomplete="off" aria-label="Search site" placeholder="Search for" data-search-index="../search.json">
</form></li>
<li class="nav-item"><a class="external-link nav-link" href="https://github.com/NicWir/VisomX/" aria-label="GitHub"><span class="fa fab fa-github fa-lg"></span></a></li>
      </ul>
</div>


  </div>
</nav><div class="container template-article">




<div class="row">
  <main id="main" class="col-md-9"><div class="page-header">

      <h1>Reliable and extensive analysis of transcriptomics data with package `VisomX`</h1>
                        <h4 data-toc-skip class="author">Nicolas T.
Wirth</h4>
            
            <h4 data-toc-skip class="date">2022-01-31</h4>
      
      <small class="dont-index">Source: <a href="https://github.com/NicWir/VisomX/blob/master/vignettes/vignette_transcriptomics.Rmd" class="external-link"><code>vignettes/vignette_transcriptomics.Rmd</code></a></small>
      <div class="d-none name"><code>vignette_transcriptomics.Rmd</code></div>
    </div>

    
    
<p style="text-align:center;">
</p>
<p><img src="https://user-images.githubusercontent.com/43075653/166063689-45b4e34e-7129-42c3-b109-3a34afdc24e2.png" class="center" width="433" height="178"></p>

<div class="section level2">
<h2 id="introduction">Introduction<a class="anchor" aria-label="anchor" href="#introduction"></a>
</h2>
<p>The aim of the VisomX’s transcriptomics section is to provide an
integrated analysis workflow for robust and reproducible analysis of
RNAseq data for differential gene expression. This section is built on
the foundation of the R package DESeq2 <span class="citation">(<a href="#ref-love2014">Love <em>et al.</em>, 2014</a>)</span>. VisomX was
designed to be user-friendly, requiring minimal prior knowledge of the R
programming language or programming in general. However, a basic
understanding of running commands from a console or writing simple
scripts is necessary. For those new to R, <a href="https://moderndive.netlify.app/1-getting-started.html" class="external-link">this</a>
website is a great starting point, there are some good resources <a href="https://education.rstudio.com/learn/beginner/" class="external-link">here</a> and I
suggest using the <a href="https://posit.co/products/open-source/rstudio/" class="external-link">RStudio
application</a>. It provides an environment for writing and running R
code.</p>
<p>With consideration for <code>R</code> novices, <code>VisomX</code>
establishes a framework in which a complete, detailed RNASeq data
analysis can be performed in two simple steps:</p>
<ol style="list-style-type: decimal">
<li><p><em>Read data</em> from one or several tabular inputs with gene
counts, assign new names and apply feature filtering.</p></li>
<li><p><em>Run workflow</em>, including missing value imputation,
normalization, log2 transformation, statistical analysis, differential
expression analysis, log2 fold change shrinkage, pathway enrichment
analysis, and rendering of a report that summarizes the
results.</p></li>
</ol>
<p>All computational results of a workflow are stored in a data
container (list) and can be visualized by passing them to a set of
dedicated plotting functions.</p>
<div style="page-break-after: always;"></div>
</div>
<div class="section level2">
<h2 id="installation">Installation<a class="anchor" aria-label="anchor" href="#installation"></a>
</h2>
<p>Install the most current version with package
<code>devtools</code>:</p>
<div class="sourceCode" id="cb1"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/utils/install.packages.html" class="external-link">install.packages</a></span><span class="op">(</span><span class="st">"devtools"</span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://devtools.r-lib.org/" class="external-link">devtools</a></span><span class="op">)</span></span>
<span></span>
<span><span class="fu">install_github</span><span class="op">(</span><span class="st">"NicWir/VisomX"</span>, dependencies <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span></span></code></pre></div>
<p>Load the package:</p>
<div class="sourceCode" id="cb2"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Load the VisomX package</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://github.com/NicWir/VisomX" class="external-link">VisomX</a></span><span class="op">)</span></span></code></pre></div>
</div>
<div class="section level2">
<h2 id="references">References<a class="anchor" aria-label="anchor" href="#references"></a>
</h2>
<div id="refs" class="references csl-bib-body hanging-indent" entry-spacing="0">
<div id="ref-love2014" class="csl-entry">
Love M.I., Huber W. &amp; Anders S. (2014). Moderated estimation of fold
change and dispersion for RNA-seq data with DESeq2. Genome Biology 15
(12): 550. <a href="https://doi.org/10.1186/s13059-014-0550-8" class="external-link">https://doi.org/10.1186/s13059-014-0550-8</a>.
</div>
</div>
</div>
  </main><aside class="col-md-3"><nav id="toc" aria-label="Table of contents"><h2>On this page</h2>
    </nav></aside>
</div>



    <footer><div class="pkgdown-footer-left">
  <p>Developed by Nicolas T. Wirth.</p>
</div>

<div class="pkgdown-footer-right">
  <p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.1.2.</p>
</div>

    </footer>
</div>





  </body>
</html>
